// @gsa-sam/components/esm5/lib/menu/menu-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Inject, Input, HostBinding, ChangeDetectionStrategy, ViewEncapsulation, Optional } from '@angular/core';
import { FocusMonitor } from '@angular/cdk/a11y';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { SDS_MENU_TOKEN } from './menu.component';
import { SdsMenuHeaderComponent } from './menu-header.component';
var SdsMenuItemComponent = /** @class */ (function () {
    function SdsMenuItemComponent(_elementRef, _focusMonitor, _parentMenu, _parentMenuHeader) {
        this._elementRef = _elementRef;
        this._focusMonitor = _focusMonitor;
        this._parentMenu = _parentMenu;
        this._parentMenuHeader = _parentMenuHeader;
        /**
         * Menu item class
         */
        this.class = this._getClass();
        /**
         * ARIA role for the menu item.
         */
        this.role = 'menuitem';
        /**
         * Menu item tab index
         */
        this.tabIndex = this.disabled ? '-1' : '0';
        /**
         * Holds the disable status value
         */
        this._disabled = false;
        // Start listening to focus changes
        _focusMonitor.monitor(this._elementRef, false);
        // Add this menu item to its parent menu
        // If item its inside a header
        // add it as the first item in the list
        if (_parentMenuHeader) {
            _parentMenu.insertItem(this, 0);
        }
        else {
            _parentMenu.addItem(this);
        }
    }
    Object.defineProperty(SdsMenuItemComponent.prototype, "disabled", {
        /** Whether the menu item should be disabled */
        get: /**
         * Whether the menu item should be disabled
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    /** Get item class */
    /**
     * Get item class
     * @return {?}
     */
    SdsMenuItemComponent.prototype._getClass = /**
     * Get item class
     * @return {?}
     */
    function () {
        /** @type {?} */
        var headerButtonSmall = this._parentMenu.size === 'sm' ? 'sds-button--small' : '';
        return this._parentMenuHeader
            ? "sds-button sds-button--circular " + headerButtonSmall
            : 'sds-menu__item';
    };
    /** Focuses the menu item. */
    /**
     * Focuses the menu item.
     * @param {?=} origin
     * @return {?}
     */
    SdsMenuItemComponent.prototype.focus = /**
     * Focuses the menu item.
     * @param {?=} origin
     * @return {?}
     */
    function (origin) {
        if (origin === void 0) { origin = 'program'; }
        this._focusMonitor.focusVia(this._elementRef.nativeElement, origin);
    };
    /** Stop listening to focus changes and remove item from parent */
    /**
     * Stop listening to focus changes and remove item from parent
     * @return {?}
     */
    SdsMenuItemComponent.prototype.ngOnDestroy = /**
     * Stop listening to focus changes and remove item from parent
     * @return {?}
     */
    function () {
        this._focusMonitor.stopMonitoring(this._elementRef);
        this._parentMenu.removeItem(this);
    };
    SdsMenuItemComponent.decorators = [
        { type: Component, args: [{
                    // tslint:disable-next-line: component-selector
                    selector: '[sds-menu-item]',
                    template: '<ng-content></ng-content>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    SdsMenuItemComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: FocusMonitor },
        { type: undefined, decorators: [{ type: Inject, args: [SDS_MENU_TOKEN,] }] },
        { type: SdsMenuHeaderComponent, decorators: [{ type: Optional }] }
    ]; };
    SdsMenuItemComponent.propDecorators = {
        class: [{ type: HostBinding, args: ['class',] }],
        role: [{ type: HostBinding, args: ['attr.role',] }, { type: Input }],
        tabIndex: [{ type: HostBinding, args: ['attr.tabIndex',] }],
        disabled: [{ type: Input }]
    };
    return SdsMenuItemComponent;
}());
export { SdsMenuItemComponent };
if (false) {
    /**
     * Menu item class
     * @type {?}
     */
    SdsMenuItemComponent.prototype.class;
    /**
     * ARIA role for the menu item.
     * @type {?}
     */
    SdsMenuItemComponent.prototype.role;
    /**
     * Menu item tab index
     * @type {?}
     */
    SdsMenuItemComponent.prototype.tabIndex;
    /**
     * Holds the disable status value
     * @type {?}
     * @private
     */
    SdsMenuItemComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    SdsMenuItemComponent.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    SdsMenuItemComponent.prototype._focusMonitor;
    /**
     * @type {?}
     * @private
     */
    SdsMenuItemComponent.prototype._parentMenu;
    /**
     * @type {?}
     * @private
     */
    SdsMenuItemComponent.prototype._parentMenuHeader;
}
