// @gsa-sam/components/esm2015/lib/menu/menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, InjectionToken, Input, Output, TemplateRef, ViewChild, ViewEncapsulation } from '@angular/core';
import { ESCAPE, DOWN_ARROW, UP_ARROW, HOME, END, hasModifierKey } from '@angular/cdk/keycodes';
import { FocusKeyManager } from '@angular/cdk/a11y';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { Subscription } from 'rxjs';
import { sdsMenuAnimations } from './menu-animations';
/**
 * Injection token used to provide the parent menu to menu items.
 * @type {?}
 */
export const SDS_MENU_TOKEN = new InjectionToken('SDS_MENU_TOKEN');
/**
 * Menu Interface
 * @record
 * @template T
 */
export function SdsMenuInterface() { }
if (false) {
    /** @type {?} */
    SdsMenuInterface.prototype.xPosition;
    /** @type {?} */
    SdsMenuInterface.prototype.yPosition;
    /** @type {?} */
    SdsMenuInterface.prototype.overlapTrigger;
    /** @type {?} */
    SdsMenuInterface.prototype.templateRef;
    /** @type {?} */
    SdsMenuInterface.prototype.closed;
    /** @type {?|undefined} */
    SdsMenuInterface.prototype.parentMenu;
    /** @type {?|undefined} */
    SdsMenuInterface.prototype.size;
    /** @type {?} */
    SdsMenuInterface.prototype.focusFirstItem;
    /** @type {?|undefined} */
    SdsMenuInterface.prototype.setPositionClasses;
    /** @type {?|undefined} */
    SdsMenuInterface.prototype.addItem;
    /** @type {?|undefined} */
    SdsMenuInterface.prototype.insertItem;
    /** @type {?|undefined} */
    SdsMenuInterface.prototype.removeItem;
}
export class SdsMenuComponent {
    /**
     * @param {?} _elementRef
     */
    constructor(_elementRef) {
        this._elementRef = _elementRef;
        /**
         * After | Before the menu triger element
         */
        this._xPosition = 'after';
        /**
         * Above | Below the menu triger element
         */
        this._yPosition = 'below';
        /**
         * Menu items inside the current menu.
         */
        this._items = [];
        /**
         * Subscription to tab events on the menu panel
         */
        this._tabSubscription = Subscription.EMPTY;
        /**
         * Config object to be passed into the menu's ngClass
         */
        this._classList = {};
        /**
         * Current state of the panel animation.
         */
        this._panelAnimationState = 'void';
        this._overlapTrigger = false;
        /**
         * Event emitted when the menu is closed.
         */
        this.closed = new EventEmitter();
    }
    /**
     * Position of the menu in the X axis.
     * @return {?}
     */
    get xPosition() {
        return this._xPosition;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set xPosition(value) {
        this._xPosition = value;
        this.setPositionClasses();
    }
    /**
     * Position of the menu in the Y axis.
     * @return {?}
     */
    get yPosition() {
        return this._yPosition;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set yPosition(value) {
        this._yPosition = value;
        this.setPositionClasses();
    }
    /**
     * Whether menu panel overlaps trigger element
     * @return {?}
     */
    get overlapTrigger() {
        return this._overlapTrigger;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set overlapTrigger(value) {
        this._overlapTrigger = coerceBooleanProperty(value);
    }
    /**
     * Transfer classes from the sds-menu to the overlay container
     * @param {?} classes
     * @return {?}
     */
    set panelClass(classes) {
        /** @type {?} */
        const previousPanelClass = this._previousPanelClass;
        // Remove previous classes from current set of classes
        if (previousPanelClass && previousPanelClass.length) {
            previousPanelClass.split(' ').forEach((/**
             * @param {?} className
             * @return {?}
             */
            (className) => {
                this._classList[className] = false;
            }));
        }
        this._previousPanelClass = classes;
        // Adds new classes to current set of classes
        if (classes && classes.length) {
            classes.split(' ').forEach((/**
             * @param {?} className
             * @return {?}
             */
            (className) => {
                this._classList[className] = true;
            }));
            // Remove all classes from <sds-menu>
            this._elementRef.nativeElement.className = '';
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setPositionClasses();
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this._keyManager = new FocusKeyManager(this._items).withWrap();
        this._tabSubscription = this._keyManager.tabOut.subscribe((/**
         * @return {?}
         */
        () => this.closed.emit('tab')));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._tabSubscription.unsubscribe();
        this.closed.complete();
    }
    /**
     * Focus the first item in the menu
     * @param {?=} origin
     * @return {?}
     */
    focusFirstItem(origin = 'program') {
        this._keyManager.setFocusOrigin(origin).setFirstItemActive();
    }
    /**
     * Adds classes to the menu panel based on its position
     * @param {?=} posX
     * @param {?=} posY
     * @return {?}
     */
    setPositionClasses(posX = this.xPosition, posY = this.yPosition) {
        /** @type {?} */
        const classes = this._classList;
        classes['sds-menu-before'] = posX === 'before';
        classes['sds-menu-after'] = posX === 'after';
        classes['sds-menu-above'] = posY === 'above';
        classes['sds-menu-below'] = posY === 'below';
    }
    /**
     * Adds a menu item with the menu.
     * @param {?} item
     * @return {?}
     */
    addItem(item) {
        if (this._items.indexOf(item) === -1) {
            this._items.push(item);
        }
    }
    /**
     * Inserts a menu item at an index
     * @param {?} item
     * @param {?} index
     * @return {?}
     */
    insertItem(item, index) {
        if (this._items.indexOf(item) === -1 && index < this._items.length) {
            this._items.splice(index, 0, item);
        }
    }
    /**
     * Removes an item from the menu.
     * @param {?} item
     * @return {?}
     */
    removeItem(item) {
        /** @type {?} */
        const index = this._items.indexOf(item);
        if (this._items.indexOf(item) > -1) {
            this._items.splice(index, 1);
        }
    }
    /**
     * Handle a keyboard event from the menu
     * @param {?} event
     * @return {?}
     */
    _handleKeydown(event) {
        // tslint:disable-next-line: deprecation
        /** @type {?} */
        const keyCode = event.keyCode;
        /** @type {?} */
        const manager = this._keyManager;
        switch (keyCode) {
            case ESCAPE:
                this.closed.emit('keydown');
                break;
            case HOME:
            case END:
                if (!hasModifierKey(event)) {
                    keyCode === HOME
                        ? manager.setFirstItemActive()
                        : manager.setLastItemActive();
                    event.preventDefault();
                }
                break;
            default:
                if (keyCode === UP_ARROW || keyCode === DOWN_ARROW) {
                    manager.setFocusOrigin('keyboard');
                }
                manager.onKeydown(event);
        }
    }
    /**
     * Starts the enter animation.
     * @return {?}
     */
    _startAnimation() {
        this._panelAnimationState = 'enter';
    }
    /**
     * Callback that is invoked when the panel animation completes.
     * @param {?} event
     * @return {?}
     */
    _onAnimationDone(event) { }
    /**
     * Resets the panel animation to its initial state.
     * @return {?}
     */
    _resetAnimation() {
        this._panelAnimationState = 'void';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    _onAnimationStart(event) {
        // Scroll the content element to the top as soon as the animation starts.
        if (event.toState === 'enter' && this._keyManager.activeItemIndex === 0) {
            event.element.scrollTop = 0;
        }
    }
}
SdsMenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-menu',
                exportAs: 'sdsMenu',
                template: "<ng-template>\n  <div\n    class=\"sds-overlay minw-menu maxw-mobile radius-overlay\"\n    [ngClass]=\"_classList\"\n    (keydown)=\"_handleKeydown($event)\"\n    (click)=\"closed.emit('click')\"\n    [@transformMenu]=\"_panelAnimationState\"\n    (@transformMenu.start)=\"_onAnimationStart($event)\"\n    (@transformMenu.done)=\"_onAnimationDone($event)\"\n    tabindex=\"-1\"\n    role=\"menu\"\n  >\n    <div class=\"sds-menu\" [class.sds-menu--small]=\"size === 'sm'\">\n      <ng-content></ng-content>\n    </div>\n  </div>\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None,
                animations: [sdsMenuAnimations.transformMenu],
                providers: [
                    { provide: SDS_MENU_TOKEN, useExisting: SdsMenuComponent }
                ]
            }] }
];
/** @nocollapse */
SdsMenuComponent.ctorParameters = () => [
    { type: ElementRef }
];
SdsMenuComponent.propDecorators = {
    templateRef: [{ type: ViewChild, args: [TemplateRef,] }],
    size: [{ type: Input }],
    xPosition: [{ type: Input }],
    yPosition: [{ type: Input }],
    overlapTrigger: [{ type: Input }],
    panelClass: [{ type: Input, args: ['class',] }],
    closed: [{ type: Output }]
};
if (false) {
    /**
     * After | Before the menu triger element
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._xPosition;
    /**
     * Above | Below the menu triger element
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._yPosition;
    /**
     * Manage browser focus
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._keyManager;
    /**
     * Menu items inside the current menu.
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._items;
    /**
     * Subscription to tab events on the menu panel
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._tabSubscription;
    /**
     * Stores <sds-menu> classes
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._previousPanelClass;
    /**
     * Config object to be passed into the menu's ngClass
     * @type {?}
     */
    SdsMenuComponent.prototype._classList;
    /**
     * Current state of the panel animation.
     * @type {?}
     */
    SdsMenuComponent.prototype._panelAnimationState;
    /**
     * Grab the component template
     * @type {?}
     */
    SdsMenuComponent.prototype.templateRef;
    /**
     * Size of menu component.
     * Affects the font-size of the menu items and
     * the size of the close button in the menu header
     * @type {?}
     */
    SdsMenuComponent.prototype.size;
    /**
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._overlapTrigger;
    /**
     * Event emitted when the menu is closed.
     * @type {?}
     */
    SdsMenuComponent.prototype.closed;
    /**
     * @type {?}
     * @private
     */
    SdsMenuComponent.prototype._elementRef;
}
