// @gsa-sam/components/esm2015/lib/menu/menu-trigger.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, Output, ViewContainerRef, HostBinding, HostListener, } from '@angular/core';
import { FocusMonitor, isFakeMousedownFromScreenReader } from '@angular/cdk/a11y';
import { Overlay, OverlayConfig } from '@angular/cdk/overlay';
import { TemplatePortal } from '@angular/cdk/portal';
import { normalizePassiveListenerOptions } from '@angular/cdk/platform';
import { merge, Subscription } from 'rxjs';
import { SdsMenuComponent } from './menu.component';
export class SdsMenuTriggerForDirective {
    /**
     * @param {?} _overlay
     * @param {?} _element
     * @param {?} _viewContainerRef
     * @param {?} _focusMonitor
     */
    constructor(_overlay, _element, _viewContainerRef, _focusMonitor) {
        this._overlay = _overlay;
        this._element = _element;
        this._viewContainerRef = _viewContainerRef;
        this._focusMonitor = _focusMonitor;
        /**
         * ARIA haspopup for the menu trigger.
         */
        this.ariaHasPopup = true;
        /**
         * Holds value for menuOpen variable
         */
        this._menuOpen = false;
        /**
         * PortalOutlet
         */
        this._overlayRef = null;
        /**
         * Tracks input type
         */
        this._openedBy = null;
        this._closingActionsSubscription = Subscription.EMPTY;
        this._menuCloseSubscription = Subscription.EMPTY;
        /**
         * Event emitted when the associated menu is opened.
         */
        this.menuOpened = new EventEmitter();
        /**
         * Event emitted when the associated menu is closed.
         */
        this.menuClosed = new EventEmitter();
        // On touch devices set _openedBy to 'touch'
        _element.nativeElement.addEventListener('touchstart', (/**
         * @return {?}
         */
        () => (this._openedBy = 'touch')), normalizePassiveListenerOptions({ passive: true }));
    }
    /**
     * ARIA expanded for the menu trigger.
     * @return {?}
     */
    get menuOpen() {
        return this._menuOpen;
    }
    /**
     * References the menu instance that the trigger is associated with.
     * @return {?}
     */
    get menu() {
        return this._menu;
    }
    /**
     * @param {?} menu
     * @return {?}
     */
    set menu(menu) {
        if (menu === this._menu) {
            return;
        }
        this._menu = menu;
        this._menuCloseSubscription.unsubscribe();
        if (menu) {
            this._menuCloseSubscription = menu.closed.asObservable().subscribe((/**
             * @return {?}
             */
            () => {
                this._destroyMenu();
            }));
        }
    }
    /**
     * Handles mouse presses on the trigger.
     * @param {?} event
     * @return {?}
     */
    _handleMousedown(event) {
        if (!isFakeMousedownFromScreenReader(event)) {
            // Since right or middle button clicks won't trigger the `click` event,
            // we shouldn't consider the menu as opened by mouse in those cases.
            this._openedBy = event.button === 0 ? 'mouse' : null;
        }
    }
    /**
     * Toggles the menu between the open and closed states.
     * @return {?}
     */
    toggleMenu() {
        return this._menuOpen ? this.closeMenu() : this.openMenu();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._overlayRef) {
            this._overlayRef.dispose();
            this._overlayRef = null;
        }
        this._element.nativeElement.removeEventListener('touchstart', (/**
         * @return {?}
         */
        () => (this._openedBy = 'touch')), normalizePassiveListenerOptions({ passive: true }));
        this._menuCloseSubscription.unsubscribe();
        this._closingActionsSubscription.unsubscribe();
    }
    /**
     * Opens the menu.
     * @return {?}
     */
    openMenu() {
        if (this._menuOpen) {
            return;
        }
        /** @type {?} */
        const overlayRef = this._createOverlay();
        /** @type {?} */
        const overlayConfig = overlayRef.getConfig();
        this._setPosition((/** @type {?} */ (overlayConfig.positionStrategy)));
        overlayRef.attach(this._getPortal());
        this._closingActionsSubscription = this._menuClosingActions().subscribe((/**
         * @return {?}
         */
        () => this.closeMenu()));
        this._initMenu();
        if (this.menu instanceof SdsMenuComponent) {
            this.menu._startAnimation();
        }
    }
    /**
     * Closes the menu.
     * @return {?}
     */
    closeMenu() {
        this.menu.closed.emit();
    }
    /**
     * Focuses the menu trigger.
     * @param {?=} origin
     * @return {?}
     */
    focus(origin = 'program') {
        this._focusMonitor.focusVia(this._element, origin);
    }
    /**
     * This method sets the menu state to open and focuses the first item
     * @private
     * @return {?}
     */
    _initMenu() {
        this._setIsMenuOpen(true);
        this.menu.focusFirstItem(this._openedBy || 'program');
    }
    /**
     * sets open state
     * @private
     * @param {?} isOpen
     * @return {?}
     */
    _setIsMenuOpen(isOpen) {
        this._menuOpen = isOpen;
        this._menuOpen ? this.menuOpened.emit() : this.menuClosed.emit();
    }
    /**
     * This method resets the menu when it's closed,
     * most importantly restoring focus to the menu trigger
     * @private
     * @return {?}
     */
    _resetMenu() {
        this._setIsMenuOpen(false);
        if (!this._openedBy) {
            this.focus();
        }
        else {
            this.focus(this._openedBy);
        }
        this._openedBy = null;
    }
    /**
     * Closes the menu and does the necessary cleanup.
     * @private
     * @return {?}
     */
    _destroyMenu() {
        if (!this._overlayRef || !this.menuOpen) {
            return;
        }
        this._closingActionsSubscription.unsubscribe();
        this._overlayRef.detach();
        if (this.menu instanceof SdsMenuComponent) {
            this.menu._resetAnimation();
            this._resetMenu();
        }
    }
    /**
     * This method creates the overlay from the provided menu's template and saves its
     * OverlayRef so that it can be attached to the DOM when openMenu is called.
     * @private
     * @return {?}
     */
    _createOverlay() {
        if (!this._overlayRef) {
            /** @type {?} */
            const config = this._getOverlayConfig();
            this._subscribeToPositions((/** @type {?} */ (config.positionStrategy)));
            this._overlayRef = this._overlay.create(config);
            this._overlayRef.keydownEvents().subscribe();
        }
        return this._overlayRef;
    }
    /**
     * This method builds the configuration object needed to create the overlay, the OverlayState.
     * @private
     * @return {?} OverlayConfig
     */
    _getOverlayConfig() {
        return new OverlayConfig({
            positionStrategy: this._overlay
                .position()
                .flexibleConnectedTo(this._element)
                .withLockedPosition()
                .withTransformOriginOn('.sds-overlay'),
            hasBackdrop: true,
            backdropClass: 'cdk-overlay-transparent-backdrop',
            scrollStrategy: this._overlay.scrollStrategies.reposition()
        });
    }
    /**
     * Listens to changes in the position of the overlay and sets the correct classes
     * on the menu based on the new position.
     * @private
     * @param {?} position
     * @return {?}
     */
    _subscribeToPositions(position) {
        position.positionChanges.subscribe((/**
         * @param {?} change
         * @return {?}
         */
        change => {
            /** @type {?} */
            const posX = change.connectionPair.overlayX === 'start' ? 'after' : 'before';
            /** @type {?} */
            const posY = change.connectionPair.overlayY === 'top' ? 'below' : 'above';
            this.menu.setPositionClasses(posX, posY);
        }));
    }
    /**
     * Sets the appropriate positions on a position strategy
     * so the overlay connects with the trigger correctly.
     * @private
     * @param {?} positionStrategy
     * @return {?}
     */
    _setPosition(positionStrategy) {
        const [originX, originFallbackX] = this.menu.xPosition === 'before' ? ['end', 'start'] : ['start', 'end'];
        const [overlayY, overlayFallbackY] = this.menu.yPosition === 'above' ? ['bottom', 'top'] : ['top', 'bottom'];
        let [originY, originFallbackY] = [overlayY, overlayFallbackY];
        const [overlayX, overlayFallbackX] = [originX, originFallbackX];
        /** @type {?} */
        const offsetY = 0;
        if (!this.menu.overlapTrigger) {
            originY = overlayY === 'top' ? 'bottom' : 'top';
            originFallbackY = overlayFallbackY === 'top' ? 'bottom' : 'top';
        }
        // Positions from most to least desirable
        positionStrategy.withPositions([
            { originX, originY, overlayX, overlayY, offsetY },
            {
                originX: originFallbackX,
                originY,
                overlayX: overlayFallbackX,
                overlayY,
                offsetY
            },
            {
                originX,
                originY: originFallbackY,
                overlayX,
                overlayY: overlayFallbackY,
                offsetY: -offsetY
            },
            {
                originX: originFallbackX,
                originY: originFallbackY,
                overlayX: overlayFallbackX,
                overlayY: overlayFallbackY,
                offsetY: -offsetY
            }
        ]);
    }
    /**
     * Returns a stream that emits whenever an action that should close the menu occurs.
     * @private
     * @return {?}
     */
    _menuClosingActions() {
        /** @type {?} */
        const backdrop = this._overlayRef.backdropClick();
        /** @type {?} */
        const detachments = this._overlayRef.detachments();
        return merge(backdrop, detachments);
    }
    /**
     * Gets the portal that should be attached to the overlay.
     * @private
     * @return {?}
     */
    _getPortal() {
        if (!this._portal || this._portal.templateRef !== this.menu.templateRef) {
            this._portal = new TemplatePortal(this.menu.templateRef, this._viewContainerRef);
        }
        return this._portal;
    }
}
SdsMenuTriggerForDirective.decorators = [
    { type: Directive, args: [{
                selector: '[sdsMenuTriggerFor]'
            },] }
];
/** @nocollapse */
SdsMenuTriggerForDirective.ctorParameters = () => [
    { type: Overlay },
    { type: ElementRef },
    { type: ViewContainerRef },
    { type: FocusMonitor }
];
SdsMenuTriggerForDirective.propDecorators = {
    ariaHasPopup: [{ type: HostBinding, args: ['attr.aria-haspopup',] }],
    menuOpen: [{ type: HostBinding, args: ['attr.aria-expanded',] }],
    menu: [{ type: Input, args: ['sdsMenuTriggerFor',] }],
    menuOpened: [{ type: Output }],
    menuClosed: [{ type: Output }],
    _handleMousedown: [{ type: HostListener, args: ['mousedown', ['$event'],] }],
    toggleMenu: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /**
     * ARIA haspopup for the menu trigger.
     * @type {?}
     */
    SdsMenuTriggerForDirective.prototype.ariaHasPopup;
    /**
     * Holds the menu instance
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._menu;
    /**
     * Holds value for menuOpen variable
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._menuOpen;
    /**
     * CDK Portal for menu panel
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._portal;
    /**
     * PortalOutlet
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._overlayRef;
    /**
     * Tracks input type
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._openedBy;
    /**
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._closingActionsSubscription;
    /**
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._menuCloseSubscription;
    /**
     * Event emitted when the associated menu is opened.
     * @type {?}
     */
    SdsMenuTriggerForDirective.prototype.menuOpened;
    /**
     * Event emitted when the associated menu is closed.
     * @type {?}
     */
    SdsMenuTriggerForDirective.prototype.menuClosed;
    /**
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._overlay;
    /**
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    SdsMenuTriggerForDirective.prototype._focusMonitor;
}
