// @gsa-sam/components/esm5/lib/observers/observe-width.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Output, EventEmitter, ElementRef } from '@angular/core';
import { ViewportRuler } from '@angular/cdk/overlay';
import { startWith } from 'rxjs/operators';
/**
 * Directive that triggers a callback whenever the width of
 * its associated element has changed.
 */
var SdsObserveWidthDirective = /** @class */ (function () {
    function SdsObserveWidthDirective(_elementRef, viewportRuler) {
        this._elementRef = _elementRef;
        this.viewportRuler = viewportRuler;
        /**
         * Event emitted for each change in the element's width.
         */
        this.elementWidth = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SdsObserveWidthDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.windowResize$ = this.viewportRuler
            .change(0)
            .pipe(startWith(this._getElementWidth()))
            .subscribe((/**
         * @return {?}
         */
        function () { return _this._emitEvent(); }));
    };
    /**
     * @return {?}
     */
    SdsObserveWidthDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.windowResize$.unsubscribe();
    };
    /**
     * @return {?}
     */
    SdsObserveWidthDirective.prototype._emitEvent = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var width = this._getElementWidth();
        this.elementWidth.emit(width);
    };
    /**
     * @return {?}
     */
    SdsObserveWidthDirective.prototype._getElementWidth = /**
     * @return {?}
     */
    function () {
        return this._elementRef.nativeElement.offsetWidth;
    };
    SdsObserveWidthDirective.decorators = [
        { type: Directive, args: [{ selector: '[sdsObserveWidth]' },] }
    ];
    /** @nocollapse */
    SdsObserveWidthDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ViewportRuler }
    ]; };
    SdsObserveWidthDirective.propDecorators = {
        elementWidth: [{ type: Output, args: ['sdsObserveWidth',] }]
    };
    return SdsObserveWidthDirective;
}());
export { SdsObserveWidthDirective };
if (false) {
    /**
     * Event emitted for each change in the element's width.
     * @type {?}
     */
    SdsObserveWidthDirective.prototype.elementWidth;
    /**
     * Subscription to window resize stream
     * @type {?}
     */
    SdsObserveWidthDirective.prototype.windowResize$;
    /**
     * @type {?}
     * @private
     */
    SdsObserveWidthDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    SdsObserveWidthDirective.prototype.viewportRuler;
}
