// @gsa-sam/components/esm5/lib/toolbar/toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { coerceBooleanProperty } from "@angular/cdk/coercion";
import { sdsToolbarAnimations } from "./toolbar-animations";
/**
 * Counter for generating unique element ids.
 * @type {?}
 */
var uniqueId = 0;
var SdsToolbarComponent = /** @class */ (function () {
    function SdsToolbarComponent() {
        /**
         * The unique Toolbar id.
         */
        this.id = uniqueId++;
        /**
         * ID for the header element. Used for a11y labelling.
         */
        this._headerId = "sds-toolbar-header-" + this.id;
        /**
         * ID for the content element. Used for a11y labelling.
         */
        this._contentId = "sds-toolbar-content-" + this.id;
        /**
         * Emits whenever the expanded state of the toolbar changes.
         */
        this.expandedChange = new EventEmitter();
        this._expanded = false;
        this._disabled = false;
        /**
         * Width of the toolbar when is expanded.
         */
        this.expandedWidth = "300px";
        // Elements that surround the toolbar need to make space for the expanded toolbar.
        // Adding 20 more pixes to provide padding
        this.expandedSpace = "20px";
    }
    Object.defineProperty(SdsToolbarComponent.prototype, "expanded", {
        /** Whether the toolbar is expanded. */
        get: /**
         * Whether the toolbar is expanded.
         * @return {?}
         */
        function () {
            return this._expanded;
        },
        set: /**
         * @param {?} expanded
         * @return {?}
         */
        function (expanded) {
            expanded = coerceBooleanProperty(expanded);
            // Only emit events and update the internal value if the value changes.
            if (this._expanded !== expanded) {
                this._expanded = expanded;
                this.expandedChange.emit(expanded);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SdsToolbarComponent.prototype, "disabled", {
        /** Whether the toolbar is disabled. */
        get: /**
         * Whether the toolbar is disabled.
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} disabled
         * @return {?}
         */
        function (disabled) {
            this._disabled = coerceBooleanProperty(disabled);
        },
        enumerable: true,
        configurable: true
    });
    /** Toggles the expanded state of the toolbar. */
    /**
     * Toggles the expanded state of the toolbar.
     * @return {?}
     */
    SdsToolbarComponent.prototype.toggle = /**
     * Toggles the expanded state of the toolbar.
     * @return {?}
     */
    function () {
        this.expanded = !this.expanded;
    };
    /** Gets the expanded state string. */
    /**
     * Gets the expanded state string.
     * @return {?}
     */
    SdsToolbarComponent.prototype._getExpandedState = /**
     * Gets the expanded state string.
     * @return {?}
     */
    function () {
        return this.expanded ? "expanded" : "collapsed";
    };
    SdsToolbarComponent.decorators = [
        { type: Component, args: [{
                    selector: "sds-toolbar",
                    exportAs: "sdsToolbar",
                    template: "<sds-toolbar-header></sds-toolbar-header>\n<div\n  class=\"sds-toolbar__content\"\n  role=\"region\"\n  [@bodyExpansion]=\"{\n    value: _getExpandedState(),\n    params: { expandedWidth: expandedWidth }\n  }\"\n  [attr.aria-labelledby]=\"_headerId\"\n  [id]=\"_contentId\"\n>\n  <ng-content></ng-content>\n</div>\n",
                    animations: [sdsToolbarAnimations.bodyExpansion],
                    host: {
                        class: "sds-toolbar",
                        "[class.sds-toolbar--expanded]": "expanded"
                    }
                }] }
    ];
    /** @nocollapse */
    SdsToolbarComponent.ctorParameters = function () { return []; };
    SdsToolbarComponent.propDecorators = {
        expandedChange: [{ type: Output }],
        expanded: [{ type: Input }],
        disabled: [{ type: Input }],
        expandedWidth: [{ type: Input }],
        expandedSpace: [{ type: Input }]
    };
    return SdsToolbarComponent;
}());
export { SdsToolbarComponent };
if (false) {
    /**
     * The unique Toolbar id.
     * @type {?}
     */
    SdsToolbarComponent.prototype.id;
    /**
     * ID for the header element. Used for a11y labelling.
     * @type {?}
     */
    SdsToolbarComponent.prototype._headerId;
    /**
     * ID for the content element. Used for a11y labelling.
     * @type {?}
     */
    SdsToolbarComponent.prototype._contentId;
    /**
     * Emits whenever the expanded state of the toolbar changes.
     * @type {?}
     */
    SdsToolbarComponent.prototype.expandedChange;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarComponent.prototype._expanded;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarComponent.prototype._disabled;
    /**
     * Width of the toolbar when is expanded.
     * @type {?}
     */
    SdsToolbarComponent.prototype.expandedWidth;
    /** @type {?} */
    SdsToolbarComponent.prototype.expandedSpace;
}
