// @gsa-sam/components/esm5/lib/toolbar/toolbar-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Host } from "@angular/core";
import { SdsToolbarComponent } from "./toolbar.component";
var SdsToolbarHeaderComponent = /** @class */ (function () {
    function SdsToolbarHeaderComponent(toolbar) {
        this.toolbar = toolbar;
    }
    /** Toggles the expanded state of the toolbar. */
    /**
     * Toggles the expanded state of the toolbar.
     * @return {?}
     */
    SdsToolbarHeaderComponent.prototype._toggle = /**
     * Toggles the expanded state of the toolbar.
     * @return {?}
     */
    function () {
        this.toolbar.toggle();
    };
    Object.defineProperty(SdsToolbarHeaderComponent.prototype, "_isExpanded", {
        /** Gets whether the toolbar is expanded. */
        get: /**
         * Gets whether the toolbar is expanded.
         * @return {?}
         */
        function () {
            return this.toolbar.expanded;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SdsToolbarHeaderComponent.prototype, "_disabled", {
        /** Whether the toolbar is disabled. */
        get: /**
         * Whether the toolbar is disabled.
         * @return {?}
         */
        function () {
            return this.toolbar.disabled;
        },
        enumerable: true,
        configurable: true
    });
    SdsToolbarHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: "sds-toolbar-header",
                    template: "<button\n  class=\"sds-toolbar__toogle-btn\"\n  [attr.id]=\"toolbar._headerId\"\n  [attr.tabindex]=\"_disabled ? -1 : 0\"\n  [attr.aria-controls]=\"toolbar._contentId\"\n  [attr.aria-expanded]=\"_isExpanded\"\n  [attr.aria-disabled]=\"_disabled\"\n  (click)=\"_toggle()\"\n>\n  <div class=\"sds-toolbar__toggle\">\n    <span class=\"usa-sr-only\">Open/Close Toolbar</span>\n  </div>\n</button>\n",
                    host: {
                        class: "sds-toolbar__header"
                    }
                }] }
    ];
    /** @nocollapse */
    SdsToolbarHeaderComponent.ctorParameters = function () { return [
        { type: SdsToolbarComponent, decorators: [{ type: Host }] }
    ]; };
    return SdsToolbarHeaderComponent;
}());
export { SdsToolbarHeaderComponent };
if (false) {
    /** @type {?} */
    SdsToolbarHeaderComponent.prototype.toolbar;
}
