// @gsa-sam/components/esm5/lib/toolbar/toolbar-expand.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2 } from "@angular/core";
import { SdsToolbarComponent } from "./toolbar.component";
var SdsToolbarExpandDirective = /** @class */ (function () {
    function SdsToolbarExpandDirective(renderer, _element) {
        this.renderer = renderer;
        this._element = _element;
    }
    Object.defineProperty(SdsToolbarExpandDirective.prototype, "toolbar", {
        /** References the toolbar instance that the element its associated with. */
        get: /**
         * References the toolbar instance that the element its associated with.
         * @return {?}
         */
        function () {
            return this._toolbar;
        },
        set: /**
         * @param {?} toolbar
         * @return {?}
         */
        function (toolbar) {
            if (toolbar === this._toolbar) {
                return;
            }
            this._toolbar = toolbar;
            /** Sets toolbar expanded width */
            this._expandedWidth = this._toolbar.expandedSpace;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SdsToolbarExpandDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.setStyle(this.toolbar.expanded);
        this.toolbar.expandedChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            _this.setStyle(value);
        }));
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    SdsToolbarExpandDirective.prototype.setStyle = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value) {
            this.renderer.setStyle(this._element.nativeElement, "margin-left", "" + this._expandedWidth);
        }
        else {
            this.renderer.removeStyle(this._element.nativeElement, "margin-left");
        }
    };
    SdsToolbarExpandDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[sdsToolbarExpand]"
                },] }
    ];
    /** @nocollapse */
    SdsToolbarExpandDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    SdsToolbarExpandDirective.propDecorators = {
        toolbar: [{ type: Input, args: ["sdsToolbarExpand",] }]
    };
    return SdsToolbarExpandDirective;
}());
export { SdsToolbarExpandDirective };
if (false) {
    /**
     * Width of the toolbar while expanded.
     * @type {?}
     */
    SdsToolbarExpandDirective.prototype._expandedWidth;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarExpandDirective.prototype._toolbar;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarExpandDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarExpandDirective.prototype._element;
}
