// @gsa-sam/components/esm5/lib/header/header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, ElementRef, HostListener } from '@angular/core';
import { NavigationHelper } from '../common-navigation/navigation-helper';
var SdsHeaderComponent = /** @class */ (function () {
    function SdsHeaderComponent() {
        this.mobileNavActive = false;
        /**
         * Navigation helper
         */
        this.navigationHelper = new NavigationHelper();
        /**
         * determines if the top banner is shown
         */
        this.showTopBanner = true;
        this.topBannerDescription = '';
        /**
         * event for event based
         */
        this.linkEvent = new EventEmitter();
    }
    /**
     * Takes in a text string and removes all white space characters and returns the new string
     * @param text
     */
    /**
     * Takes in a text string and removes all white space characters and returns the new string
     * @param {?} text
     * @return {?}
     */
    SdsHeaderComponent.prototype.removeWhiteSpace = /**
     * Takes in a text string and removes all white space characters and returns the new string
     * @param {?} text
     * @return {?}
     */
    function (text) {
        return text.replace(/ /g, '');
    };
    /**
     * seeif any secondary link has a counter
     */
    /**
     * seeif any secondary link has a counter
     * @return {?}
     */
    SdsHeaderComponent.prototype.hasCounter = /**
     * seeif any secondary link has a counter
     * @return {?}
     */
    function () {
        /** @type {?} */
        var hasCounter = false;
        if (this.model) {
            if (this.model.secondaryLinks) {
                this.model.secondaryLinks.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    if (item.hasCounter) {
                        hasCounter = true;
                    }
                }));
            }
        }
        return hasCounter;
    };
    /**
     * Deselects previous seletion
     * @param id
     */
    /**
     * Deselects previous seletion
     * @param {?} id
     * @return {?}
     */
    SdsHeaderComponent.prototype.select = /**
     * Deselects previous seletion
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.deselect();
        /** @type {?} */
        var item = this.find(id);
        if (item) {
            item.selected = true;
        }
    };
    /**
     * Deselects all the items in the header model
     */
    /**
     * Deselects all the items in the header model
     * @return {?}
     */
    SdsHeaderComponent.prototype.deselect = /**
     * Deselects all the items in the header model
     * @return {?}
     */
    function () {
        if (this.model) {
            if (this.model.home) {
                this.model.home.selected = false;
            }
            if (this.model.navigationLinks) {
                this.model.navigationLinks.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    item.selected = false;
                    if (item.children) {
                        item.children.forEach((/**
                         * @param {?} child
                         * @return {?}
                         */
                        function (child) {
                            child.selected = false;
                        }));
                    }
                }));
            }
            if (this.model.secondaryLinks) {
                this.model.secondaryLinks.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    item.selected = false;
                }));
            }
        }
    };
    /**
     * Finds the navigation element by id in the header model
     * @param id of the navigation item
     */
    /**
     * Finds the navigation element by id in the header model
     * @param {?} id of the navigation item
     * @return {?}
     */
    SdsHeaderComponent.prototype.find = /**
     * Finds the navigation element by id in the header model
     * @param {?} id of the navigation item
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var toReturn = null;
        if (this.model) {
            if (this.model.home) {
                if (this.model.home.id === id) {
                    toReturn = this.model.home;
                }
            }
            toReturn = this.findNavigationLinks(id, toReturn);
            if (this.model.secondaryLinks) {
                this.model.secondaryLinks.forEach((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) {
                    if (item.id === id) {
                        toReturn = item;
                    }
                }));
            }
        }
        return toReturn;
    };
    /**
     * Searchs the items in the navigation links
     * @param id
     */
    /**
     * Searchs the items in the navigation links
     * @private
     * @param {?} id
     * @param {?} toReturn
     * @return {?}
     */
    SdsHeaderComponent.prototype.findNavigationLinks = /**
     * Searchs the items in the navigation links
     * @private
     * @param {?} id
     * @param {?} toReturn
     * @return {?}
     */
    function (id, toReturn) {
        if (this.model.navigationLinks) {
            this.model.navigationLinks.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                if (item.id === id) {
                    toReturn = item;
                }
                if (item.children) {
                    item.children.forEach((/**
                     * @param {?} child
                     * @return {?}
                     */
                    function (child) {
                        if (child.id === id) {
                            toReturn = child;
                        }
                    }));
                }
            }));
        }
        return toReturn;
    };
    /**
     * Link clicked and emits the link data into an event
     * @param link
     */
    /**
     * Link clicked and emits the link data into an event
     * @param {?} link
     * @return {?}
     */
    SdsHeaderComponent.prototype.linkClickEvent = /**
     * Link clicked and emits the link data into an event
     * @param {?} link
     * @return {?}
     */
    function (link) {
        this.linkEvent.emit(link);
        return false;
    };
    // When the mobile nav is active, and the close box isn't visible,
    // we know the user's viewport has been resized to be larger.
    // Let's make the page state consistent by deactivating the mobile nav.
    // When the mobile nav is active, and the close box isn't visible,
    // we know the user's viewport has been resized to be larger.
    // Let's make the page state consistent by deactivating the mobile nav.
    /**
     * @param {?} event
     * @return {?}
     */
    SdsHeaderComponent.prototype.onBrowserResize = 
    // When the mobile nav is active, and the close box isn't visible,
    // we know the user's viewport has been resized to be larger.
    // Let's make the page state consistent by deactivating the mobile nav.
    /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.mobileNavActive &&
            this.closeNavBtn.nativeElement.getBoundingClientRect().width === 0) {
            this.mobileNavActive = false;
        }
    };
    /**
     * @return {?}
     */
    SdsHeaderComponent.prototype.openMobileNav = /**
     * @return {?}
     */
    function () {
        this.mobileNavActive = true;
    };
    /**
     * @return {?}
     */
    SdsHeaderComponent.prototype.closeMobileNav = /**
     * @return {?}
     */
    function () {
        this.mobileNavActive = false;
        // The mobile nav was just deactivated, and focus was on the close
        // button, which is no longer visible. We don't want the focus to
        // disappear into the void, so focus on the menu button if it's
        // visible (this may have been what the user was just focused on,
        // if they triggered the mobile nav by mistake).
        this.openNavBtn.nativeElement.focus();
    };
    // The mobile nav was just activated, so focus on the close button,
    // The mobile nav was just activated, so focus on the close button,
    /**
     * @return {?}
     */
    SdsHeaderComponent.prototype.navAnimationEnd = 
    // The mobile nav was just activated, so focus on the close button,
    /**
     * @return {?}
     */
    function () {
        this.closeNavBtn.nativeElement.focus();
    };
    SdsHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-header',
                    template: "<sds-top-banner *ngIf=\"showTopBanner\" \n[description]=\"topBannerDescription\"\n></sds-top-banner>\n<div\n  class=\"usa-overlay\"\n  [class.is-visible]=\"mobileNavActive\"\n  (click)=\"closeMobileNav()\"\n></div>\n<header class=\"usa-header usa-header--extended\">\n  <div class=\"usa-navbar\">\n    <div class=\"usa-logo\" id=\"extended-logo\">\n      <ng-container\n        [ngTemplateOutlet]=\"\n          navigationHelper.isLinkInternal(model.home)\n            ? homeLinkRouteTemplate\n            : homeLinkHREFTemplate\n        \"\n        [ngTemplateOutletContext]=\"{ $implicit: model.home }\"\n      ></ng-container>\n    </div>\n    <button #usaNavOpen class=\"usa-menu-btn\" (click)=\"openMobileNav()\">\n      <fa-layers [fixedWidth]=\"true\" size=\"3x\">\n        <fa-icon\n          [icon]=\"['fas', 'square']\"\n          [classes]=\"['text-primary']\"\n        ></fa-icon>\n        <fa-icon\n          [icon]=\"['sds', 'bars']\"\n          [classes]=\"['icon-bars']\"\n          transform=\"shrink-6\"\n        ></fa-icon>\n        <fa-layers-counter\n          *ngIf=\"hasCounter()\"\n          [classes]=\"['icon-layers-counter']\"\n        ></fa-layers-counter>\n      </fa-layers>\n      <span class=\"usa-sr-only\">Menu</span>\n    </button>\n  </div>\n  <nav\n    aria-label=\"Primary navigation\"\n    class=\"usa-nav\"\n    (keydown.esc)=\"closeMobileNav()\"\n    (animationend)=\"navAnimationEnd()\"\n    [class.is-visible]=\"mobileNavActive\"\n    [cdkTrapFocus]=\"mobileNavActive\"\n  >\n    <div class=\"usa-nav__inner\">\n      <button #usaNavClose class=\"usa-nav__close\" (click)=\"closeMobileNav()\">\n        <fa-layers [fixedWidth]=\"true\" size=\"lg\">\n          <fa-icon\n            [icon]=\"['fas', 'circle']\"\n            [inverse]=\"true\"\n            transform=\"grow-2\"\n          ></fa-icon>\n          <fa-icon [icon]=\"['fas', 'times']\" transform=\"shrink-6\"></fa-icon>\n        </fa-layers>\n        <span class=\"usa-sr-only\">Close</span>\n      </button>\n      <ul class=\"usa-nav__primary usa-accordion\">\n        <li\n          *ngFor=\"let link of model.navigationLinks\"\n          class=\"usa-nav__primary-item\"\n        >\n          <ng-container\n            [ngTemplateOutlet]=\"link.route ? linkTemplate : dropDownTemplate\"\n            [ngTemplateOutletContext]=\"{ $implicit: link }\"\n          ></ng-container>\n        </li>\n      </ul>\n      <div class=\"usa-nav__secondary\">\n        <ul class=\"usa-nav__secondary-links\">\n          <li\n            *ngFor=\"let link of model.secondaryLinks\"\n            class=\"usa-nav__secondary-item\"\n          >\n            <ng-container\n              [ngTemplateOutlet]=\"\n                navigationHelper.isLinkInternal(link)\n                  ? secondaryRouteLinkTemplate\n                  : navigationHelper.isLinkExternal(link)\n                  ? secondaryHREFLinkTemplate\n                  : secondaryEVENTLinkTemplate\n              \"\n              [ngTemplateOutletContext]=\"{ $implicit: link }\"\n            ></ng-container>\n          </li>\n        </ul>\n      </div>\n    </div>\n  </nav>\n</header>\n<ng-template #homeLinkRouteTemplate let-home>\n  <a [routerLink]=\"home.route\" title=\"Home\" aria-label=\"Home\">\n    <img class=\"sds-header__logo\" [src]=\"home.logo\" [alt]=\"home.text\" />\n  </a>\n</ng-template>\n\n<ng-template #homeLinkHREFTemplate let-home>\n  <a [attr.href]=\"home.route\" title=\"Home\" aria-label=\"Home\">\n    <img class=\"sds-header__logo\" [src]=\"home.logo\" [alt]=\"home.text\" />\n  </a>\n</ng-template>\n\n<ng-template #secondaryRouteLinkTemplate let-link>\n  <a\n    [attr.id]=\"link.id\"\n    [routerLink]=\"[link.route]\"\n    [attr.class]=\"link.selected ? 'usa-current' : ''\"\n  >\n    <ng-container\n      [ngTemplateOutlet]=\"iconLinkTemplate\"\n      [ngTemplateOutletContext]=\"{ $implicit: link }\"\n    >\n    </ng-container>\n  </a>\n</ng-template>\n\n<ng-template #secondaryHREFLinkTemplate let-link>\n  <a\n    [attr.id]=\"link.id\"\n    [attr.href]=\"[link.route]\"\n    [attr.class]=\"link.selected ? 'usa-current' : ''\"\n  >\n    <ng-container\n      [ngTemplateOutlet]=\"iconLinkTemplate\"\n      [ngTemplateOutletContext]=\"{ $implicit: link }\"\n    >\n    </ng-container>\n  </a>\n</ng-template>\n\n<ng-template #secondaryEVENTLinkTemplate let-link>\n  <a\n    [attr.id]=\"link.id\"\n    (click)=\"linkClickEvent(link)\"\n    href=\"javascript:void(0)\"\n    [attr.class]=\"link.selected ? 'usa-current' : ''\"\n  >\n    <ng-container\n      [ngTemplateOutlet]=\"iconLinkTemplate\"\n      [ngTemplateOutletContext]=\"{ $implicit: link }\"\n    >\n    </ng-container>\n  </a>\n</ng-template>\n\n<ng-template #iconLinkTemplate let-link>\n  <fa-layers [fixedWidth]=\"true\">\n    <fa-icon [icon]=\"[link.imageClassPrefix, link.imageClass]\"></fa-icon>\n    <fa-layers-counter\n      *ngIf=\"link.hasCounter\"\n      [classes]=\"['icon-layers-counter']\"\n    ></fa-layers-counter>\n  </fa-layers>\n  <span class=\"sds-nav__secondary-item-text\">{{ link.text }}</span>\n</ng-template>\n\n<ng-template #linkTemplate let-link>\n  <ng-container\n    [ngTemplateOutlet]=\"\n      navigationHelper.isLinkInternal(link)\n        ? linkRouteTemplate\n        : navigationHelper.isLinkExternal(link)\n        ? linkHREFTemplate\n        : linkEventTemplate\n    \"\n    [ngTemplateOutletContext]=\"{ $implicit: link }\"\n  ></ng-container>\n</ng-template>\n\n<ng-template #linkRouteTemplate let-link>\n  <a\n    [attr.id]=\"link.id\"\n    [attr.class]=\"link.selected ? 'usa-nav__link usa-current' : 'usa-nav__link'\"\n    [routerLink]=\"[link.route]\"\n    ><span>{{ link.text }}</span></a\n  >\n</ng-template>\n\n<ng-template #linkHREFTemplate let-link>\n  <a\n    [attr.id]=\"link.id\"\n    [attr.class]=\"link.selected ? 'usa-nav__link usa-current' : 'usa-nav__link'\"\n    [attr.href]=\"[link.route]\"\n    ><span>{{ link.text }}</span></a\n  >\n</ng-template>\n\n<ng-template #linkEventTemplate let-link>\n  <a\n    [attr.id]=\"link.id\"\n    [attr.class]=\"link.selected ? 'usa-nav__link usa-current' : 'usa-nav__link'\"\n    href=\"javascript:void(0)\"\n    (click)=\"linkClickEvent(link)\"\n    ><span>{{ link.text }}</span></a\n  >\n</ng-template>\n\n<ng-template #dropDownTemplate let-link>\n  <button\n    [attr.id]=\"link.id\"\n    [attr.class]=\"\n      link.selected\n        ? 'usa-accordion__button usa-nav__link usa-current'\n        : 'usa-accordion__button usa-nav__link'\n    \"\n    aria-expanded=\"false\"\n    [attr.aria-controls]=\"removeWhiteSpace(link.text) + '-section'\"\n  >\n    <span>{{ link.text }}</span>\n  </button>\n  <ul\n    [attr.id]=\"removeWhiteSpace(link.text) + '-section'\"\n    class=\"usa-nav__submenu\"\n    hidden\n  >\n    <li *ngFor=\"let childLink of link.children\" class=\"usa-nav__submenu-item\">\n      <ng-container\n        [ngTemplateOutlet]=\"\n          navigationHelper.isLinkInternal(childLink)\n            ? dropDownRouteLinkTemplate\n            : navigationHelper.isLinkExternal(childLink)\n            ? dropDownHREFLinkTemplate\n            : dropDownEVENTLinkTemplate\n        \"\n        [ngTemplateOutletContext]=\"{ $implicit: childLink }\"\n      ></ng-container>\n    </li>\n  </ul>\n</ng-template>\n\n<ng-template #dropDownRouteLinkTemplate let-childLink>\n  <a [routerLink]=\"[childLink.route]\">{{ childLink.text }}</a>\n</ng-template>\n\n<ng-template #dropDownHREFLinkTemplate let-childLink>\n  <a [attr.href]=\"[childLink.route]\">{{ childLink.text }}</a>\n</ng-template>\n\n<ng-template #dropDownEVENTLinkTemplate let-childLink>\n  <a href=\"javascript:void(0)\" (click)=\"linkClickEvent(childLink)\">{{\n    childLink.text\n  }}</a>\n</ng-template>\n",
                    styles: [""]
                }] }
    ];
    SdsHeaderComponent.propDecorators = {
        openNavBtn: [{ type: ViewChild, args: ['usaNavOpen',] }],
        closeNavBtn: [{ type: ViewChild, args: ['usaNavClose',] }],
        showTopBanner: [{ type: Input }],
        model: [{ type: Input }],
        topBannerDescription: [{ type: Input }],
        linkEvent: [{ type: Output }],
        onBrowserResize: [{ type: HostListener, args: ['window:resize', ['$event'],] }]
    };
    return SdsHeaderComponent;
}());
export { SdsHeaderComponent };
if (false) {
    /** @type {?} */
    SdsHeaderComponent.prototype.openNavBtn;
    /** @type {?} */
    SdsHeaderComponent.prototype.closeNavBtn;
    /** @type {?} */
    SdsHeaderComponent.prototype.mobileNavActive;
    /**
     * Navigation helper
     * @type {?}
     */
    SdsHeaderComponent.prototype.navigationHelper;
    /**
     * determines if the top banner is shown
     * @type {?}
     */
    SdsHeaderComponent.prototype.showTopBanner;
    /**
     * Model used for the different display portions of the header
     * @type {?}
     */
    SdsHeaderComponent.prototype.model;
    /** @type {?} */
    SdsHeaderComponent.prototype.topBannerDescription;
    /**
     * event for event based
     * @type {?}
     */
    SdsHeaderComponent.prototype.linkEvent;
}
