// @gsa-sam/sam-formly/esm5/lib/formly-filters/advanced-filters/advanced-filters.component.js
import { Component, Input } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { SdsDialogService } from '@gsa-sam/components';
import { SdsAdvancedFiltersService } from './sds-advanced-filters.service';
import { SdsFormlyDialogComponent } from '../../formly-dialog/formly-dialog.component';
var AdvancedFiltersComponent = /** @class */ (function () {
    function AdvancedFiltersComponent(dialog, advancedFiltersService) {
        this.dialog = dialog;
        this.advancedFiltersService = advancedFiltersService;
        /**
         *    Options for the form.
         */
        this.options = {};
    }
    AdvancedFiltersComponent.prototype.openDialog = function () {
        var _this = this;
        var modalFields = this.advancedFiltersService.convertToCheckboxes(this.fields);
        var data = {
            fields: modalFields,
            originalFields: this.fields,
            originalModel: this.model,
            submit: 'Update',
            title: 'More Filters',
        };
        var dialogRef = this.dialog.open(SdsFormlyDialogComponent, {
            width: 'medium',
            data: data
        });
        dialogRef.afterClosed().subscribe(function (result) {
            if (result) {
                _this.fields = result.fields;
                _this.model = result.model;
            }
        });
    };
    AdvancedFiltersComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-advanced-filters',
                    template: "<button type=\"button\" (click)=\"openDialog()\" class=\"usa-button usa-button--unstyled\">More\n  Filters\n  <fa-icon [icon]=\"['sds', 'filter']\" [classes]=\"['icon-filter']\"></fa-icon>\n</button>\n",
                    styles: [".usa-button--unstyled{padding-left:0;width:initial;text-decoration:none;font-weight:600}.usa-button--unstyled:active,.usa-button--unstyled:hover{background:initial;color:initial;text-decoration:underline}"]
                }] }
    ];
    /** @nocollapse */
    AdvancedFiltersComponent.ctorParameters = function () { return [
        { type: SdsDialogService },
        { type: SdsAdvancedFiltersService }
    ]; };
    AdvancedFiltersComponent.propDecorators = {
        form: [{ type: Input }],
        fields: [{ type: Input }],
        model: [{ type: Input }],
        options: [{ type: Input }]
    };
    return AdvancedFiltersComponent;
}());
export { AdvancedFiltersComponent };
