// @gsa-sam/components/esm5/lib/truncate-text/truncate.text.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { SdsTruncateTextByLineDirective } from './truncate-text.directive';
import { SdsTruncatedTextContainerComponent } from './truncate-text-container.component';
var SdsTruncateModule = /** @class */ (function () {
    function SdsTruncateModule() {
    }
    SdsTruncateModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    exports: [SdsTruncateTextByLineDirective],
                    declarations: [
                        SdsTruncateTextByLineDirective,
                        SdsTruncatedTextContainerComponent
                    ],
                    providers: [],
                    entryComponents: [SdsTruncatedTextContainerComponent]
                },] }
    ];
    return SdsTruncateModule;
}());
export { SdsTruncateModule };
