// @gsa-sam/components/esm5/lib/truncate-text/truncate-text-container.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, HostBinding, HostListener } from '@angular/core';
import { SDS_TRUNCATED_TEXT_DATA } from './truncates-text-base';
import { sdsTruncateTextAnimations } from './truncate-text-animations';
/**
 * @record
 */
export function SdsTruncateTextData() { }
if (false) {
    /** @type {?} */
    SdsTruncateTextData.prototype.text;
}
var SdsTruncatedTextContainerComponent = /** @class */ (function () {
    function SdsTruncatedTextContainerComponent(data) {
        this.data = data;
        this._animationState = 'void';
    }
    /** Starts the animation. */
    /**
     * Starts the animation.
     * @return {?}
     */
    SdsTruncatedTextContainerComponent.prototype.startAnimation = /**
     * Starts the animation.
     * @return {?}
     */
    function () {
        this._animationState = 'enter';
    };
    /** Resets the animation to its initial state. */
    /**
     * Resets the animation to its initial state.
     * @return {?}
     */
    SdsTruncatedTextContainerComponent.prototype.resetAnimation = /**
     * Resets the animation to its initial state.
     * @return {?}
     */
    function () {
        this._animationState = 'void';
    };
    /** Intentionally left empty to trigger change detection */
    /**
     * Intentionally left empty to trigger change detection
     * @return {?}
     */
    SdsTruncatedTextContainerComponent.prototype._onAnimationDone = /**
     * Intentionally left empty to trigger change detection
     * @return {?}
     */
    function () { };
    SdsTruncatedTextContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-truncated-text-container',
                    template: "\n    <div class=\"sds-overlay maxw-mobile radius-overlay padding-2\">{{ data.text }}</div>\n  ",
                    animations: [sdsTruncateTextAnimations.container]
                }] }
    ];
    /** @nocollapse */
    SdsTruncatedTextContainerComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [SDS_TRUNCATED_TEXT_DATA,] }] }
    ]; };
    SdsTruncatedTextContainerComponent.propDecorators = {
        _animationState: [{ type: HostBinding, args: ['@container',] }],
        _onAnimationDone: [{ type: HostListener, args: ['@container.done',] }]
    };
    return SdsTruncatedTextContainerComponent;
}());
export { SdsTruncatedTextContainerComponent };
if (false) {
    /** @type {?} */
    SdsTruncatedTextContainerComponent.prototype._animationState;
    /** @type {?} */
    SdsTruncatedTextContainerComponent.prototype.data;
}
