// @gsa-sam/components/esm2015/lib/observers/observe-width.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Output, EventEmitter, ElementRef } from '@angular/core';
import { ViewportRuler } from '@angular/cdk/overlay';
import { startWith } from 'rxjs/operators';
/**
 * Directive that triggers a callback whenever the width of
 * its associated element has changed.
 */
export class SdsObserveWidthDirective {
    /**
     * @param {?} _elementRef
     * @param {?} viewportRuler
     */
    constructor(_elementRef, viewportRuler) {
        this._elementRef = _elementRef;
        this.viewportRuler = viewportRuler;
        /**
         * Event emitted for each change in the element's width.
         */
        this.elementWidth = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.windowResize$ = this.viewportRuler
            .change(0)
            .pipe(startWith(this._getElementWidth()))
            .subscribe((/**
         * @return {?}
         */
        () => this._emitEvent()));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.windowResize$.unsubscribe();
    }
    /**
     * @return {?}
     */
    _emitEvent() {
        /** @type {?} */
        const width = this._getElementWidth();
        this.elementWidth.emit(width);
    }
    /**
     * @return {?}
     */
    _getElementWidth() {
        return this._elementRef.nativeElement.offsetWidth;
    }
}
SdsObserveWidthDirective.decorators = [
    { type: Directive, args: [{ selector: '[sdsObserveWidth]' },] }
];
/** @nocollapse */
SdsObserveWidthDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ViewportRuler }
];
SdsObserveWidthDirective.propDecorators = {
    elementWidth: [{ type: Output, args: ['sdsObserveWidth',] }]
};
if (false) {
    /**
     * Event emitted for each change in the element's width.
     * @type {?}
     */
    SdsObserveWidthDirective.prototype.elementWidth;
    /**
     * Subscription to window resize stream
     * @type {?}
     */
    SdsObserveWidthDirective.prototype.windowResize$;
    /**
     * @type {?}
     * @private
     */
    SdsObserveWidthDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    SdsObserveWidthDirective.prototype.viewportRuler;
}
