// @gsa-sam/sam-formly/bundles/gsa-sam-sam-formly.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/material/input'), require('@angular/material/core'), require('@angular/material/datepicker'), require('@angular/animations'), require('@ngx-formly/core/select'), require('@fortawesome/angular-fontawesome'), require('@angular/router'), require('qs'), require('ts-md5/dist/md5'), require('rxjs'), require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('@ngx-formly/core'), require('@gsa-sam/components')) :
    typeof define === 'function' && define.amd ? define('@gsa-sam/sam-formly', ['exports', '@angular/material/input', '@angular/material/core', '@angular/material/datepicker', '@angular/animations', '@ngx-formly/core/select', '@fortawesome/angular-fontawesome', '@angular/router', 'qs', 'ts-md5/dist/md5', 'rxjs', '@angular/core', '@angular/common', '@angular/forms', '@ngx-formly/core', '@gsa-sam/components'], factory) :
    (factory((global['gsa-sam'] = global['gsa-sam'] || {}, global['gsa-sam']['sam-formly'] = {}),global.ng.material.input,global.ng.material.core,global.ng.material.datepicker,global.ng.animations,global.select,global.angularFontawesome,global.ng.router,global.qs,global.md5,global.rxjs,global.ng.core,global.ng.common,global.ng.forms,global.core$1,global.components));
}(this, (function (exports,input,core,datepicker,animations,select,angularFontawesome,router,qs,md5,rxjs,i0,common,forms,core$1,components) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    var FormlyCustomWrapperComponent = /** @class */ (function (_super) {
        __extends(FormlyCustomWrapperComponent, _super);
        function FormlyCustomWrapperComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyCustomWrapperComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sam-formly-filter-wrapper-form-field',
                        template: "\n  <sds-accordion multi=\"true\" displayMode=\"basic\">\n    <sds-accordion-item class=\"sds-accordion__panel\">\n      <sds-accordion-item-header> {{to.label}} </sds-accordion-item-header>\n      <div class=\"usa-form-group\" [class.usa-form-group--error]=\"showError\">\n      <label class=\"usa-label\" *ngIf=\"to.label && to.hideLabel !== true\" [attr.for]=\"id\" [ngClass]=\"to.labelClass\">\n        <span *ngIf=\"to.tagText\" class=\"usa-tag\"  [ngClass]=\"to.tagClass ? to.tagClass : 'sds-tag--info-white'\">{{to.tagText}}</span> \n        <span>{{ to.label }}</span>\n        <span *ngIf=\"!to.required && !to.hideOptional\"> (Optional)</span>\n      </label>  \n      <small *ngIf=\"to.description\" class=\"form-text text-muted\">{{ to.description }}</small>\n      <ng-template #fieldComponent></ng-template>\n      <div *ngIf=\"showError\" class=\"usa-error-message\" [style.display]=\"'block'\">\n        <formly-validation-message [field]=\"field\"></formly-validation-message>\n      </div>\n      </div>\n    </sds-accordion-item>\n  </sds-accordion>\n  "
                    }] }
        ];
        FormlyCustomWrapperComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyCustomWrapperComponent;
    }(core$1.FieldWrapper));

    /**
     * @param {string} [to.tagClass] Class to be added to the tag (default: sds-tag--info-white)
     * @param {string} [to.tagText] Text to be shown inside the tag
     * @param {string} [to.labelClass] Class to be applied to the label
     * @param {string} [to.label] Text to be shown for the label
     * @param {string} [to.required] Makes the field required
     * @param {string} [to.description] Add a description below the label
     * @param {string} [to.hideOptional] Remove the optional text
     * @param {string} [to.hideLabel] Hide the label
     *
     */
    var FormlyWrapperFormFieldComponent = /** @class */ (function (_super) {
        __extends(FormlyWrapperFormFieldComponent, _super);
        function FormlyWrapperFormFieldComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyWrapperFormFieldComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-wrapper-form-field',
                        template: "\n<div class=\"usa-form-group\" [class.usa-form-group--error]=\"showError\">\n  <label class=\"usa-label\" *ngIf=\"to.label && to.hideLabel !== true\" [attr.for]=\"id\" [ngClass]=\"to.labelClass\">\n    <span *ngIf=\"to.tagText\" class=\"usa-tag\"  [ngClass]=\"to.tagClass ? to.tagClass : 'sds-tag--info-white'\">{{to.tagText}}</span>\n    <span>{{ to.label }}</span>\n    <span *ngIf=\"!to.required && !to.hideOptional\"> (Optional)</span>\n  </label>\n  <small *ngIf=\"to.description\" class=\"form-text text-muted\">{{ to.description }}</small>\n  <ng-template #fieldComponent></ng-template>\n  <div *ngIf=\"showError\" class=\"usa-error-message\" [style.display]=\"'block'\">\n    <formly-validation-message [field]=\"field\"></formly-validation-message>\n  </div>\n</div>\n  "
                    }] }
        ];
        FormlyWrapperFormFieldComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyWrapperFormFieldComponent;
    }(core$1.FieldWrapper));

    /**
     * @param {string} [to.expand] to expand the accordion
     *
     */
    var FormlyAccordianFormFieldComponent = /** @class */ (function (_super) {
        __extends(FormlyAccordianFormFieldComponent, _super);
        function FormlyAccordianFormFieldComponent() {
            return _super.call(this) || this;
        }
        FormlyAccordianFormFieldComponent.prototype.modelHasValue = function () {
            if (this.to.hasOwnProperty('expand')) {
                return this.to.expand;
            }
            else {
                var hasValue = this.formControl.value instanceof Object
                    ? qs.stringify(this.formControl.value, { skipNulls: true })
                    : this.formControl.value;
                return hasValue ? true : false;
            }
        };
        FormlyAccordianFormFieldComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sam-formly-accordian-form-field',
                        template: "\n    <sds-accordion multi=\"true\" displayMode=\"basic\">\n      <sds-accordion-item\n        class=\"sds-accordion__panel\"\n        [expanded]=\"modelHasValue()\"\n      >\n        <sds-accordion-item-header> {{ to.label }} </sds-accordion-item-header>\n        <ng-container #fieldComponent></ng-container>\n      </sds-accordion-item>\n    </sds-accordion>\n  "
                    }] }
        ];
        /** @nocollapse */
        FormlyAccordianFormFieldComponent.ctorParameters = function () { return []; };
        FormlyAccordianFormFieldComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyAccordianFormFieldComponent;
    }(core$1.FieldWrapper));

    /**
     * @param {string} [to.ariaHidden] Hide the label
     * @param {string} [to.label] Text to be shown for the label
     */
    var FormlyFormFieldFilterWrapperComponent = /** @class */ (function (_super) {
        __extends(FormlyFormFieldFilterWrapperComponent, _super);
        function FormlyFormFieldFilterWrapperComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyFormFieldFilterWrapperComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sam-formly-filter-wrapper-form-field',
                        template: "\n    <div  class=\"wrapper-body\">\n      <div class=\"sds-accordion__trigger header-label\" [attr.aria-hidden]=\"to.ariaHidden ? 'false' : 'true'\"> {{to.label}} </div>\n      <ng-container #fieldComponent></ng-container>\n    </div>\n  "
                    }] }
        ];
        FormlyFormFieldFilterWrapperComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyFormFieldFilterWrapperComponent;
    }(core$1.FieldWrapper));

    var FormlyFieldInputComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldInputComponent, _super);
        function FormlyFieldInputComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyFieldInputComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-input',
                        template: "\n  <input [class.usa-input--error]=\"showError\" class=\"usa-input\" [placeholder]=\"to.placeholder\" name=\"input-success\" [formlyAttributes]=\"field\" [type]=\"to.inputType? to.inputType : 'text'\" [formControl]=\"formControl\">\n  "
                    }] }
        ];
        return FormlyFieldInputComponent;
    }(core$1.FieldType));

    var FormlyFieldSelectComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldSelectComponent, _super);
        function FormlyFieldSelectComponent() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.defaultOptions = {
                templateOptions: { options: [] },
            };
            return _this;
        }
        FormlyFieldSelectComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-select',
                        template: "\n    <select *ngIf=\"to.multiple; else singleSelect\" \n      multiple\n      [class.custom-select]=\"to.customSelect\"\n      [formControl]=\"formControl\"\n      [class.is-invalid]=\"showError\"\n      [formlyAttributes]=\"field\">\n        <ng-container *ngFor=\"let item of to.options | formlySelectOptions:field | async\">\n         <optgroup *ngIf=\"item.group\" label=\"{{item.label}}\">\n            <option *ngFor=\"let child of item.group\" [ngValue]=\"child.value\" [disabled]=\"child.disabled\">\n              {{ child.label }}\n            </option>\n          </optgroup>\n          <option *ngIf=\"!item.group\" [ngValue]=\"item.value\" [disabled]=\"item.disabled\">{{ item.label }}</option>\n        </ng-container>\n    </select>\n    <ng-template #singleSelect>\n      <select class=\"usa-select\"\n        [formControl]=\"formControl\"\n        [class.custom-select]=\"to.customSelect\"\n        [class.is-invalid]=\"showError\"\n        [formlyAttributes]=\"field\">\n        <option *ngIf=\"to.placeholder\" [ngValue]=\"null\">{{ to.placeholder }}</option>\n        <ng-container *ngFor=\"let item of to.options | formlySelectOptions:field | async\">\n          <optgroup *ngIf=\"item.group\" label=\"{{item.label}}\">\n            <option *ngFor=\"let child of item.group\" [ngValue]=\"child.value\" [disabled]=\"child.disabled\">\n              {{ child.label }}\n            </option>\n          </optgroup>\n          <option *ngIf=\"!item.group\" [ngValue]=\"item.value\" [disabled]=\"item.disabled\">{{ item.label }}</option>\n        </ng-container>\n      </select>\n    </ng-template>\n  "
                    }] }
        ];
        return FormlyFieldSelectComponent;
    }(core$1.FieldType));

    var FormlyFieldRadioComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldRadioComponent, _super);
        function FormlyFieldRadioComponent() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.defaultOptions = {
                templateOptions: {
                    options: []
                },
            };
            return _this;
        }
        FormlyFieldRadioComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-radio',
                        template: "\n  <div class=\"usa-radio\">\n      <div *ngFor=\"let option of to.options | formlySelectOptions:field | async; let i = index;\">\n        <input type=\"radio\"\n          [id]=\"id + '_' + i\"\n          class=\"usa-radio__input\"\n          [name]=\"id\"\n          [class.usa-input--error]=\"showError\"\n          [attr.value]=\"option.value\"\n          [value]=\"option.value\"\n          [formControl]=\"formControl\"\n          [formlyAttributes]=\"field\">\n        <label\n          class=\"usa-radio__label\"\n          [for]=\"id + '_' + i\">\n          {{ option.label }}\n        </label>\n      </div>\n    </div>\n  "
                    }] }
        ];
        return FormlyFieldRadioComponent;
    }(core$1.FieldType));

    var FormlyFieldCheckboxComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldCheckboxComponent, _super);
        function FormlyFieldCheckboxComponent() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.defaultOptions = {
                templateOptions: {
                    indeterminate: true,
                    hideLabel: true,
                },
            };
            return _this;
        }
        FormlyFieldCheckboxComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-checkbox',
                        template: "\n    <div class=\"usa-checkbox\">\n      <input class=\"usa-checkbox__input\" type=\"checkbox\"\n        [class.is-invalid]=\"showError\"\n        [indeterminate]=\"to.indeterminate && formControl.value === null\"\n        [formControl]=\"formControl\"\n        [formlyAttributes]=\"field\">\n      <label class=\"usa-checkbox__label\" [for]=\"id\">\n        {{ to.label }}\n        <span *ngIf=\"!to.required && !to.hideOptional\"> (Optional)</span>\n      </label>\n    </div>\n  "
                    }] }
        ];
        return FormlyFieldCheckboxComponent;
    }(core$1.FieldType));

    var FormlyFieldMultiCheckboxComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldMultiCheckboxComponent, _super);
        function FormlyFieldMultiCheckboxComponent() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.defaultOptions = {
                templateOptions: {
                    options: [],
                },
            };
            return _this;
        }
        FormlyFieldMultiCheckboxComponent.prototype.onChange = function (value, checked) {
            var _a;
            if (this.to.type === 'array') {
                this.formControl.patchValue(checked
                    ? __spread((this.formControl.value || []), [value]) : __spread((this.formControl.value || [])).filter(function (o) { return o !== value; }));
            }
            else {
                this.formControl.patchValue(__assign({}, this.formControl.value, (_a = {}, _a[value] = checked, _a)));
            }
            this.formControl.markAsTouched();
        };
        FormlyFieldMultiCheckboxComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-multicheckbox',
                        template: "\n  <div class=\"usa-checkbox\">\n    <div *ngFor=\"let option of to.options | formlySelectOptions:field | async; let i = index;\">\n    <input type=\"checkbox\"\n     [id]=\"id + '_' + i\"\n     class=\"usa-checkbox__input\"\n      [value]=\"option.value\"\n      [checked]=\"formControl.value && (this.to.type === 'array' ? formControl.value.includes(option.value) : formControl.value[option.value])\"\n      [formlyAttributes]=\"field\"\n      (change)=\"onChange(option.value, $event.target.checked)\">\n    <label\n    class=\"usa-checkbox__label\"\n    [for]=\"id + '_' + i\">\n        {{ option.label }}\n    </label>\n  </div>\n</div>\n  "
                    }] }
        ];
        return FormlyFieldMultiCheckboxComponent;
    }(core$1.FieldType));

    var FormlyFieldTextAreaComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldTextAreaComponent, _super);
        function FormlyFieldTextAreaComponent() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.defaultOptions = {
                templateOptions: {
                    cols: 1,
                    rows: 1,
                },
            };
            return _this;
        }
        FormlyFieldTextAreaComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-textarea',
                        template: "\n    <textarea [formControl]=\"formControl\" [cols]=\"to.cols\"\n      [rows]=\"to.rows\" class=\"usa-textarea\" [class.usa-input--error]=\"showError\"\n      [formlyAttributes]=\"field\">\n    </textarea>\n  "
                    }] }
        ];
        return FormlyFieldTextAreaComponent;
    }(core$1.FieldType));

    var AbstractSdsFormly = /** @class */ (function (_super) {
        __extends(AbstractSdsFormly, _super);
        function AbstractSdsFormly() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        AbstractSdsFormly.prototype.ngDoCheck = function () {
            this.setProperties(this.template, this.field.templateOptions);
        };
        AbstractSdsFormly.prototype.setProperties = function (component, configuration) {
            Object.keys(configuration).forEach(function (key) {
                component[key] = configuration[key];
            });
            if (this.template.control) {
                this.template.control = this.formControl;
            }
            this.cdr.detectChanges();
        };
        return AbstractSdsFormly;
    }(core$1.FieldType));

    var FormlyFieldAutoCompleteComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldAutoCompleteComponent, _super);
        function FormlyFieldAutoCompleteComponent(_cdr) {
            var _this = _super.call(this) || this;
            _this.cdr = _cdr;
            return _this;
        }
        FormlyFieldAutoCompleteComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-autocomplete',
                        template: "\n  <sds-autocomplete [formControl]=\"formControl\"></sds-autocomplete>\n  ",
                        changeDetection: i0.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        FormlyFieldAutoCompleteComponent.ctorParameters = function () {
            return [
                { type: i0.ChangeDetectorRef }
            ];
        };
        FormlyFieldAutoCompleteComponent.propDecorators = {
            template: [{ type: i0.ViewChild, args: [components.SDSAutocompleteComponent,] }]
        };
        return FormlyFieldAutoCompleteComponent;
    }(AbstractSdsFormly));

    var FormlyFieldDatePickerComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldDatePickerComponent, _super);
        function FormlyFieldDatePickerComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyFieldDatePickerComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-datepicker',
                        template: "\n    <input\n      [id]=\"id\"\n      class=\"usa-input display-inline-block margin-top-3\"\n      [class.usa-input--error]=\"showError\"\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n      matInput\n      [min]=\"to.minDate\"\n      [max]=\"to.maxDate\"\n      [matDatepicker]=\"picker\"\n      placeholder=\"Choose a date\"\n    />\n    <mat-datepicker-toggle matSuffix [for]=\"picker\"></mat-datepicker-toggle>\n    <mat-datepicker [startAt]=\"to.startDate\" #picker></mat-datepicker>\n  "
                    }] }
        ];
        return FormlyFieldDatePickerComponent;
    }(core$1.FieldType));

    var FormlyFieldButtonComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldButtonComponent, _super);
        function FormlyFieldButtonComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyFieldButtonComponent.prototype.onClick = function ($event) {
            if (this.to.onClick) {
                this.to.onClick($event);
            }
        };
        FormlyFieldButtonComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-button',
                        template: "\n      <button [type]=\"to.type\" [ngClass]=\"'btn btn-' + to.btnType\" (click)=\"onClick($event)\" class=\"usa-button--unstyled\">\n        {{ to.text }}\n      </button>\n  "
                    }] }
        ];
        return FormlyFieldButtonComponent;
    }(core$1.FieldType));

    var FormlyFieldTextComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldTextComponent, _super);
        function FormlyFieldTextComponent(_cdr) {
            var _this = _super.call(this) || this;
            _this.cdr = _cdr;
            return _this;
        }
        FormlyFieldTextComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-text',
                        template: "\n  <sds-text [formControl]=\"formControl\"></sds-text>\n  ",
                        changeDetection: i0.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        FormlyFieldTextComponent.ctorParameters = function () {
            return [
                { type: i0.ChangeDetectorRef }
            ];
        };
        FormlyFieldTextComponent.propDecorators = {
            template: [{ type: i0.ViewChild, args: [components.SdsTextComponent,] }]
        };
        return FormlyFieldTextComponent;
    }(AbstractSdsFormly));

    /**
     *
     * @param control
     * @param field
     */
    function minDateValidator(control, field) {
        var toReturn = null;
        var minDateField = field.templateOptions.minDate;
        var value = control.value;
        if (value && minDateField) {
            if (value instanceof Date && minDateField instanceof Date) {
                if (value < minDateField) {
                    if (!field.templateOptions.maxDate && !(field.templateOptions.maxDate instanceof Date)) {
                        toReturn = {
                            'minDate': true
                        };
                    }
                    else {
                        toReturn = {
                            'betweenDate': true
                        };
                    }
                }
            }
        }
        return toReturn;
    }
    /**
     *
     * @param control
     * @param field
     */
    function maxDateValidator(control, field) {
        var toReturn = null;
        var maxDateField = field.templateOptions.maxDate;
        var value = control.value;
        if (value && maxDateField) {
            if (value instanceof Date && maxDateField instanceof Date) {
                if (value > maxDateField) {
                    if (!field.templateOptions.minDate && !(field.templateOptions.minDate instanceof Date)) {
                        toReturn = {
                            'maxDate': true
                        };
                    }
                    else {
                        toReturn = {
                            'betweenDate': true
                        };
                    }
                }
            }
        }
        return toReturn;
    }

    var sdsFieldWrapper = ['label', 'description', 'validation'];
    var sdsGroupWrapper = ['group'];
    var sdsWrappers = __spread(sdsGroupWrapper, sdsFieldWrapper);

    /**
     * @param {string} [to.tagClass] Class to be added to the tag (default: sds-tag--info-white)
     * @param {string} [to.tagText] Text to be shown inside the tag
     * @param {string} [to.label] Text to be shown for the label
     * @param {string} [to.required] Makes the field required
     * @param {string} [to.hideOptional] Remove the optional text
     *
     */
    var FormlyLabelWrapperComponent = /** @class */ (function (_super) {
        __extends(FormlyLabelWrapperComponent, _super);
        function FormlyLabelWrapperComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyLabelWrapperComponent.prototype.hasLabel = function () {
            if (this.to.label) {
                if (!(this.field.type === 'checkbox' || this.field.type === 'multicheckbox')) {
                    return true;
                }
            }
        };
        FormlyLabelWrapperComponent.decorators = [
            { type: i0.Component, args: [{
                        template: "\n    <div class=\"usa-form-group\" [class.usa-form-group--error]=\"showError\">\n      <label class=\"usa-label\" *ngIf=\"hasLabel()\" [attr.for]=\"id\" [ngClass]=\"{'usa-sr-only' : to.hideLabel || ((to.group==='panel' || to.group==='accordion') && field?.parent?.type!==\n      'formly-group') }\">\n        <span *ngIf=\"to.tagText\" class=\"usa-tag\"  [ngClass]=\"to.tagClass ? to.tagClass : 'sds-tag--info-white'\">{{to.tagText}}</span>\n        <span>{{ to.label }}</span>\n        <span *ngIf=\"!to.required && !to.hideOptional\"> (Optional)</span>\n      </label>\n      <ng-container #fieldComponent></ng-container>\n    </div>\n  "
                    }] }
        ];
        FormlyLabelWrapperComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyLabelWrapperComponent;
    }(core$1.FieldWrapper));

    /**
     * @param {string} [to.description] Add a description below the label
     *
     */
    var FormlyDescriptionWrapperComponent = /** @class */ (function (_super) {
        __extends(FormlyDescriptionWrapperComponent, _super);
        function FormlyDescriptionWrapperComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyDescriptionWrapperComponent.decorators = [
            { type: i0.Component, args: [{
                        template: "\n    <div>\n      <small *ngIf=\"to.description\" class=\"form-text text-muted\">{{ to.description }}</small>\n      <ng-container #fieldComponent></ng-container>\n    </div>\n  "
                    }] }
        ];
        FormlyDescriptionWrapperComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyDescriptionWrapperComponent;
    }(core$1.FieldWrapper));

    /**
     * @param {string} [to.required] Makes the field required
     */
    var FormlyValidationWrapperComponent = /** @class */ (function (_super) {
        __extends(FormlyValidationWrapperComponent, _super);
        function FormlyValidationWrapperComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyValidationWrapperComponent.decorators = [
            { type: i0.Component, args: [{
                        template: "\n    <ng-container #fieldComponent></ng-container>\n    <div *ngIf=\"showError\" class=\"usa-error-message\" [style.display]=\"'block'\">\n      <formly-validation-message [field]=\"field\"></formly-validation-message>\n    </div>\n  "
                    }] }
        ];
        FormlyValidationWrapperComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyValidationWrapperComponent;
    }(core$1.FieldWrapper));

    /**
     * @param {string} [to.group] used to set the wrapper tupe
     * @param {string} [to.announceLabel] For screenreader
     * @param {string} [to.label] Text to be shown for the label
     * @param {string} [to.hideLabel] Hide the label
     *
     */
    var FormlyGroupWrapperComponent = /** @class */ (function (_super) {
        __extends(FormlyGroupWrapperComponent, _super);
        function FormlyGroupWrapperComponent() {
            return _super.call(this) || this;
        }
        FormlyGroupWrapperComponent.prototype.modelHasValue = function () {
            if (this.to.hasOwnProperty('expand')) {
                return this.to.expand;
            }
            else {
                var hasValue = this.formControl.value instanceof Object
                    ? qs.stringify(this.formControl.value, { skipNulls: true })
                    : this.formControl.value;
                return hasValue ? true : false;
            }
        };
        FormlyGroupWrapperComponent.decorators = [
            { type: i0.Component, args: [{
                        template: "\n    <ng-container [ngSwitch]=\"to.group\">\n      <ng-container *ngSwitchCase=\"'accordion'\">\n        <sds-accordion multi=\"true\" displayMode=\"basic\">\n          <sds-accordion-item\n            class=\"sds-accordion__panel\"\n            [expanded]=\"modelHasValue()\"\n          >\n            <sds-accordion-item-header>\n              <span\n                *ngIf=\"!to.hideLabel\"\n                [attr.aria-hidden]=\"!to.announceLabel ? undefined : 'true'\"\n              >\n                {{ to.label }}\n              </span>\n            </sds-accordion-item-header>\n            <ng-container #fieldComponent></ng-container>\n          </sds-accordion-item>\n        </sds-accordion>\n      </ng-container>\n      <ng-container *ngSwitchCase=\"'panel'\">\n        <div\n          class=\"sds-panel\"\n          [ngClass]=\"{ 'sds-panel--multiple': field?.fieldGroup?.length }\"\n        >\n          <div\n            class=\"sds-panel__header\"\n            *ngIf=\"!to.hideLabel\"\n            [attr.aria-hidden]=\"!to.announceLabel ? undefined : 'true'\"\n          >\n            {{ to.label }}\n          </div>\n          <div class=\"sds-panel__body\">\n            <ng-container #fieldComponent></ng-container>\n          </div>\n        </div>\n      </ng-container>\n      <ng-container *ngSwitchDefault>\n        <ng-container #fieldComponent></ng-container>\n      </ng-container>\n    </ng-container>\n  "
                    }] }
        ];
        /** @nocollapse */
        FormlyGroupWrapperComponent.ctorParameters = function () { return []; };
        FormlyGroupWrapperComponent.propDecorators = {
            fieldComponent: [{ type: i0.ViewChild, args: ['fieldComponent', { read: i0.ViewContainerRef },] }]
        };
        return FormlyGroupWrapperComponent;
    }(core$1.FieldWrapper));

    var FormlyFieldSearchComponent = /** @class */ (function (_super) {
        __extends(FormlyFieldSearchComponent, _super);
        function FormlyFieldSearchComponent(_cdr) {
            var _this = _super.call(this) || this;
            _this.cdr = _cdr;
            return _this;
        }
        FormlyFieldSearchComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-field-search',
                        template: "\n    <sds-search [formControl]=\"formControl\"></sds-search>\n    ",
                        changeDetection: i0.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        FormlyFieldSearchComponent.ctorParameters = function () {
            return [
                { type: i0.ChangeDetectorRef }
            ];
        };
        FormlyFieldSearchComponent.propDecorators = {
            template: [{ type: i0.ViewChild, args: [components.SdsSearchComponent,] }]
        };
        return FormlyFieldSearchComponent;
    }(AbstractSdsFormly));

    var FIELD_TYPE_COMPONENTS = [
        FormlyFieldInputComponent,
        FormlyFieldCheckboxComponent,
        FormlyFieldRadioComponent,
        FormlyFieldSelectComponent,
        FormlyFieldTextAreaComponent,
        FormlyFieldMultiCheckboxComponent,
        FormlyWrapperFormFieldComponent,
        FormlyAccordianFormFieldComponent,
        FormlyFieldAutoCompleteComponent,
        FormlyFieldDatePickerComponent,
        FormlyFormFieldFilterWrapperComponent,
        FormlyFieldButtonComponent,
        FormlyCustomWrapperComponent,
        FormlyLabelWrapperComponent,
        FormlyDescriptionWrapperComponent,
        FormlyValidationWrapperComponent,
        FormlyFieldTextComponent,
        FormlyGroupWrapperComponent,
        FormlyFieldSearchComponent
    ];
    var FORMLY_WRAPPERS = [
        {
            name: 'form-field',
            component: FormlyWrapperFormFieldComponent,
            componentName: 'FormlyWrapperFormFieldComponent'
        },
        {
            name: 'accordionwrapper',
            component: FormlyAccordianFormFieldComponent,
            componentName: 'FormlyAccordianFormFieldComponent'
        },
        {
            name: 'filterwrapper',
            component: FormlyFormFieldFilterWrapperComponent,
            componentName: 'FormlyFormFieldFilterWrapperComponent'
        },
        {
            name: 'label',
            component: FormlyLabelWrapperComponent,
            componentName: 'FormlyLabelWrapperComponent'
        },
        {
            name: 'description',
            component: FormlyDescriptionWrapperComponent,
            componentName: 'FormlyDescriptionWrapperComponent'
        },
        {
            name: 'validation',
            component: FormlyValidationWrapperComponent,
            componentName: 'FormlyValidationWrapperComponent'
        },
        {
            name: 'group',
            component: FormlyGroupWrapperComponent,
            componentName: 'FormlyGroupWrapperComponent'
        },
        {
            name: 'customwrapper',
            component: FormlyCustomWrapperComponent,
            componentName: 'FormlyCustomWrapperComponent'
        }
    ];
    var FORMLY_CONFIG = {
        types: [
            {
                name: 'formly-group',
                wrappers: sdsGroupWrapper
            },
            {
                name: 'button',
                component: FormlyFieldButtonComponent
            },
            {
                name: 'customtext',
                component: FormlyFieldTextComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'input',
                component: FormlyFieldInputComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'checkbox',
                component: FormlyFieldCheckboxComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'radio',
                component: FormlyFieldRadioComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'select',
                component: FormlyFieldSelectComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'textarea',
                component: FormlyFieldTextAreaComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'multicheckbox',
                component: FormlyFieldMultiCheckboxComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'autocomplete',
                component: FormlyFieldAutoCompleteComponent,
                wrappers: sdsWrappers
            },
            {
                name: 'datepicker',
                component: FormlyFieldDatePickerComponent,
                wrappers: sdsWrappers,
                defaultOptions: {
                    validators: {
                        validation: [maxDateValidator, minDateValidator]
                    }
                }
            },
            {
                name: 'daterangepicker',
                extends: 'formly-group',
                wrappers: sdsWrappers,
                defaultOptions: {
                    fieldGroup: [
                        {
                            type: 'datepicker',
                            key: 'fromDate',
                            templateOptions: {
                                label: 'From'
                            },
                            expressionProperties: {
                                'templateOptions.minDate': minDateFromDateRangePicker,
                                'templateOptions.maxDate': maxDateFromDateRangePicker
                            }
                        },
                        {
                            type: 'datepicker',
                            key: 'toDate',
                            templateOptions: {
                                label: 'To'
                            },
                            expressionProperties: {
                                'templateOptions.minDate': minDateToDateRangePicker,
                                'templateOptions.maxDate': maxDateToDateRangePicker
                            }
                        }
                    ]
                }
            },
            {
                name: 'search',
                component: FormlyFieldSearchComponent,
                wrappers: sdsWrappers
            }
        ],
        wrappers: [
            { name: 'label', component: FormlyLabelWrapperComponent },
            { name: 'description', component: FormlyDescriptionWrapperComponent },
            { name: 'validation', component: FormlyValidationWrapperComponent },
            { name: 'group', component: FormlyGroupWrapperComponent },
            { name: 'form-field', component: FormlyWrapperFormFieldComponent },
            { name: 'accordionwrapper', component: FormlyAccordianFormFieldComponent },
            { name: 'filterwrapper', component: FormlyFormFieldFilterWrapperComponent },
            { name: 'customwrapper', component: FormlyCustomWrapperComponent }
        ]
    };
    function minDateToDateRangePicker(model, formState, field) {
        var date = null;
        //Setting a minumn date for the date range picker
        if (field.parent.templateOptions.minDate) {
            date = new Date(field.parent.templateOptions.minDate.getTime());
        }
        if (model) {
            if (model.fromDate) {
                date = model.fromDate;
            }
        }
        return date;
    }
    function minDateFromDateRangePicker(model, formState, field) {
        var date = null;
        //Setting a minumn date for the date range picker
        if (field.parent.templateOptions.minDate) {
            date = new Date(field.parent.templateOptions.minDate.getTime());
        }
        return date;
    }
    function maxDateToDateRangePicker(model, formState, field) {
        var date = null;
        //Setting a max date for the date range picker
        if (field.parent.templateOptions.maxDate) {
            date = new Date(field.parent.templateOptions.maxDate.getTime());
        }
        return date;
    }
    function maxDateFromDateRangePicker(model, formState, field) {
        var date = null;
        //Setting a max date for the date range picker
        if (field.parent.templateOptions.maxDate) {
            date = new Date(field.parent.templateOptions.maxDate);
        }
        if (model) {
            if (model.toDate) {
                date = model.toDate;
            }
        }
        return date;
    }

    var SlideInOutAnimation = [
        animations.trigger('slideInOut', [
            animations.state('in', animations.style({
                'max-height': '2000px', 'opacity': '1', 'visibility': 'visible',
            })),
            animations.state('out', animations.style({
                'max-height': '0px', 'opacity': '0', 'visibility': 'hidden',
            })),
            animations.transition('in => out', [animations.group([
                    animations.animate('400ms ease-in-out', animations.style({
                        'opacity': '0',
                    })),
                    animations.animate('600ms ease-in-out', animations.style({
                        'max-height': '0px',
                    })),
                    animations.animate('700ms ease-in-out', animations.style({
                        'visibility': 'hidden',
                    })),
                ])]),
            animations.transition('out => in', [animations.group([
                    animations.animate('1ms ease-in-out', animations.style({
                        'visibility': 'visible',
                    })),
                    animations.animate('600ms ease-in-out', animations.style({
                        'max-height': '500px',
                    })),
                    animations.animate('800ms ease-in-out', animations.style({
                        'opacity': '1',
                    })),
                ])]),
        ]),
    ];
    var AnimationWrapperComponent = /** @class */ (function (_super) {
        __extends(AnimationWrapperComponent, _super);
        function AnimationWrapperComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        AnimationWrapperComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sam-formly-wrapper-animation',
                        template: "\n    <div [@slideInOut]=\"field.hide ? 'out' : 'in'\">\n      <ng-container #fieldComponent></ng-container>\n    </div>\n  ",
                        // styles: [`
                        //   ::ng-deep formly-field {
                        //     display: block !important;
                        //   }
                        // `],
                        animations: [SlideInOutAnimation]
                    }] }
        ];
        return AnimationWrapperComponent;
    }(core$1.FieldWrapper));

    // Validate the min length of the character
    function minlengthValidationMessage(err, field) {
        return "Should have atleast " + field.templateOptions.minLength + " characters";
    }
    // Validate the max length of the character
    function maxlengthValidationMessage(err, field) {
        return "This value should be less than " + field.templateOptions.maxLength + " characters";
    }
    // Validate the min value of the character
    function minValidationMessage(err, field) {
        return "This value should be more than " + field.templateOptions.min;
    }
    function minDateValidationMessage(err, field) {
        var dt = field.templateOptions.minDate;
        var dateFormat = (dt.getMonth() + 1) + "/" + dt.getDate() + "/" + dt.getFullYear();
        return "Date must be after " + dateFormat;
    }
    function maxDateValidationMessage(err, field) {
        var dt = field.templateOptions.maxDate;
        var dateFormat = (dt.getMonth() + 1) + "/" + dt.getDate() + "/" + dt.getFullYear();
        return "Date must be before " + dateFormat;
    }
    function betweenDateValidationMessage(err, field) {
        var dtnmax = field.templateOptions.maxDate;
        var dateMaxFormat = (dtnmax.getMonth() + 1) + "/" + dtnmax.getDate() + "/" + dtnmax.getFullYear();
        var dtmin = field.templateOptions.minDate;
        var dateMinFormat = (dtmin.getMonth() + 1) + "/" + dtmin.getDate() + "/" + dtmin.getFullYear();
        return "Date must be between " + dateMinFormat + " and " + dateMaxFormat + " ";
    }
    function invalidDateFormatValidationMessage(err, field) {
        return "Valid date format required (ex: MM/DD/YYYY)";
    }
    // Validate the max value of the character
    function maxValidationMessage(err, field) {
        return "This value should be less than " + field.templateOptions.max;
    }
    function animationExtension(field) {
        if (field.wrappers && field.wrappers.includes('animation')) {
            return;
        }
        field.wrappers = __spread(['animation'], (field.wrappers || []));
    }
    var SdsFormlyModule = /** @class */ (function () {
        function SdsFormlyModule() {
        }
        SdsFormlyModule.forChild = function (config) {
            if (config === void 0) {
                config = {};
            }
            return [{ ngModule: SdsFormlyModule, providers: [] }, core$1.FormlyModule.forChild(config)];
        };
        SdsFormlyModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [
                            FIELD_TYPE_COMPONENTS,
                            AnimationWrapperComponent,
                        ],
                        imports: [
                            common.CommonModule,
                            components.SdsAccordionModule,
                            components.SdsAutocompleteModule,
                            components.SdsDialogModule,
                            components.SdsSearchModule,
                            components.SdsTextModule,
                            core.MatNativeDateModule,
                            datepicker.MatDatepickerModule,
                            input.MatInputModule,
                            angularFontawesome.FontAwesomeModule,
                            forms.ReactiveFormsModule,
                            select.FormlySelectModule,
                            core$1.FormlyModule.forChild(FORMLY_CONFIG),
                            core$1.FormlyModule.forRoot({
                                validationMessages: [
                                    { name: 'required', message: 'This field is required' },
                                    { name: 'minlength', message: minlengthValidationMessage },
                                    { name: 'maxlength', message: maxlengthValidationMessage },
                                    { name: 'min', message: minValidationMessage },
                                    { name: 'max', message: maxValidationMessage },
                                    { name: 'minDate', message: minDateValidationMessage },
                                    { name: 'maxDate', message: maxDateValidationMessage },
                                    { name: 'betweenDate', message: betweenDateValidationMessage },
                                    { name: 'matDatepickerParse', message: invalidDateFormatValidationMessage } // Comes from the datepicker
                                ],
                                validators: [
                                    { name: 'minDate', validation: minDateValidator },
                                    { name: 'maxDate', validation: maxDateValidator }
                                ],
                                wrappers: [
                                    { name: 'animation', component: AnimationWrapperComponent },
                                ],
                                extensions: [
                                    { name: 'animation', extension: { onPopulate: animationExtension } },
                                ],
                            })
                        ]
                    },] }
        ];
        return SdsFormlyModule;
    }());

    var SdsFormlyResetComponent = /** @class */ (function () {
        function SdsFormlyResetComponent() {
            /**
             * Pass in classes for reset button -- default .usa-button .usa-button--unstyled
             */
            this.classes = ['usa-button', 'usa-button--unstyled'];
        }
        SdsFormlyResetComponent.prototype.resetAll = function () {
            this.options.resetModel();
        };
        SdsFormlyResetComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-reset',
                        template: "<button (click)=\"resetAll()\" [ngClass]=\"classes\">\n  Reset All\n  <fa-icon [icon]=\"['sds', 'reset']\" [classes]=\"['icon-reset']\"></fa-icon>\n</button>\n",
                        styles: [".usa-button--unstyled{padding-right:0;width:initial;text-decoration:none;font-weight:600}.usa-button--unstyled:active,.usa-button--unstyled:hover{background:initial;color:initial;text-decoration:underline}"]
                    }] }
        ];
        SdsFormlyResetComponent.propDecorators = {
            options: [{ type: i0.Input }],
            classes: [{ type: i0.Input }]
        };
        return SdsFormlyResetComponent;
    }());

    var SdsFormlyResetModule = /** @class */ (function () {
        function SdsFormlyResetModule() {
        }
        SdsFormlyResetModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [SdsFormlyResetComponent],
                        imports: [
                            common.CommonModule,
                            angularFontawesome.FontAwesomeModule
                        ],
                        exports: [
                            SdsFormlyResetComponent
                        ]
                    },] }
        ];
        return SdsFormlyResetModule;
    }());

    var SDSFormlyUpdateComunicationService = /** @class */ (function () {
        function SDSFormlyUpdateComunicationService() {
            /**
             * Filter update subject
             */
            this.updateFilterSource = new rxjs.Subject();
            /**
             * Observable to subscribe to for get updates of the filter
             */
            this.filterUpdate = this.updateFilterSource.asObservable();
        }
        /**
         * Update the filter
         * @param filterData
         */
        SDSFormlyUpdateComunicationService.prototype.updateFilter = function (filterData) {
            this.updateFilterSource.next(filterData);
        };
        return SDSFormlyUpdateComunicationService;
    }());

    var SdsFiltersComponent = /** @class */ (function () {
        function SdsFiltersComponent(formlyUpdateComunicationService, cdr, router$$1, route) {
            var _this = this;
            this.formlyUpdateComunicationService = formlyUpdateComunicationService;
            this.cdr = cdr;
            this.router = router$$1;
            this.route = route;
            /**
             *    Options for the form.
             */
            this.options = {};
            /**
             *  Emit results when model updated
             * To enable History Tracking
             *  If advanced filters dialog should be displayed -- defaults to false
             */
            this.advancedFilters = false;
            /**
             * Timer id for the timer awaiting the service call for more typeing
             */
            this.isHistoryEnable = true;
            /**
             *  Emit results when model updated
             */
            // TODO: check type -- Formly models are typically objects
            this.filterChange = new i0.EventEmitter();
            this.sdsFilterHistory = [];
            this._isObj = function (obj) { return typeof obj === 'object' && obj !== null; };
            this._isEmpty = function (obj) { return Object.keys(obj).length === 0; };
            this.overwrite = function (baseObj, newObj) {
                var result = {};
                for (var key in baseObj) {
                    if (Array.isArray(baseObj[key])) {
                        result[key] = newObj[key];
                    }
                    else if (_this._isObj(baseObj[key])) {
                        result[key] = _this.overwrite(baseObj[key], newObj[key] || {});
                    }
                    else {
                        result[key] = newObj[key] || null;
                    }
                }
                return result;
            };
            this.nullify = function (obj) {
                for (var key in obj) {
                    if (_this._isObj(obj[key])) {
                        obj[key] = _this.nullify(obj[key]);
                    }
                    else {
                        obj[key] = null;
                    }
                }
                return obj;
            };
        }
        SdsFiltersComponent.prototype.onpopstate = function () {
            var queryString = window.location.search;
            var urlParams = new URLSearchParams(queryString);
            var ref = urlParams.get('ref');
            var updatedFormValue = ref == null
                ? this.nullify(this.form.value)
                : JSON.parse(localStorage.getItem(ref));
            var updatedValue = this.overwrite(this.form.getRawValue(), updatedFormValue);
            this.form.setValue(updatedValue, { emitEvent: false });
            this.updateChange(updatedFormValue);
        };
        SdsFiltersComponent.prototype.ngOnInit = function () {
            var _this = this;
            if (this.isHistoryEnable) {
                var queryString = window.location.search;
                var urlParams = new URLSearchParams(queryString);
                var initialRef = urlParams.get('ref');
                if (initialRef) {
                    var updatedFormValue_1 = JSON.parse(localStorage.getItem(initialRef));
                    setTimeout(function () {
                        _this.model = __assign({}, _this.model, updatedFormValue_1);
                        _this.updateChange(updatedFormValue_1);
                        _this.cdr.detectChanges();
                    }, 0);
                }
                else {
                    this.updateChange(this.model);
                    this.clearStorage();
                }
            }
            this.cdr.detectChanges();
        };
        SdsFiltersComponent.prototype.onModelChange = function (change) {
            if (this.isHistoryEnable) {
                var md5$$1 = new md5.Md5();
                var hashCode = md5$$1.appendStr(qs.stringify(change)).end();
                this.router.navigate([], {
                    relativeTo: this.route,
                    queryParams: { ref: hashCode },
                    queryParamsHandling: 'merge'
                });
                this.addToStorageList(hashCode);
                localStorage.setItem(hashCode.toString(), JSON.stringify(change));
            }
            this.updateChange(change);
        };
        SdsFiltersComponent.prototype.updateChange = function (change) {
            this.filterChange.emit(change);
            if (this.formlyUpdateComunicationService) {
                this.formlyUpdateComunicationService.updateFilter(change);
            }
        };
        SdsFiltersComponent.prototype.addToStorageList = function (hashCode) {
            var list = JSON.parse(localStorage.getItem('sdsFilterHistory'));
            this.sdsFilterHistory = (list && list.length > 0) ? list : this.sdsFilterHistory;
            this.sdsFilterHistory.push(hashCode);
            localStorage.setItem('sdsFilterHistory', JSON.stringify(this.sdsFilterHistory));
        };
        SdsFiltersComponent.prototype.clearStorage = function () {
            var list = JSON.parse(localStorage.getItem('sdsFilterHistory'));
            if (list && list.length > 0) {
                var unique = list.filter(function (item, i, ar) { return ar.indexOf(item) === i; });
                unique.forEach(function (item) {
                    localStorage.removeItem(item);
                });
            }
        };
        SdsFiltersComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-filters',
                        template: "<formly-form [form]=\"form\" [fields]=\"fields\" [options]=\"options\" [model]=\"model\" (modelChange)=\"onModelChange($event)\">\n</formly-form>\n<div class=\"grid-row\">\n  <div *ngIf=\"advancedFilters\" class=\"grid-col\">\n    <sds-advanced-filters [form]=\"form\" [fields]=\"fields\" [options]=\"options\" [model]=\"model\">\n    </sds-advanced-filters>\n  </div>\n  <div class=\"grid-col text-right\">\n    <sds-formly-reset [options]=\"options\"></sds-formly-reset>\n  </div>\n</div>\n"
                    }] }
        ];
        /** @nocollapse */
        SdsFiltersComponent.ctorParameters = function () {
            return [
                { type: SDSFormlyUpdateComunicationService, decorators: [{ type: i0.Optional }] },
                { type: i0.ChangeDetectorRef },
                { type: router.Router },
                { type: router.ActivatedRoute }
            ];
        };
        SdsFiltersComponent.propDecorators = {
            form: [{ type: i0.Input }],
            fields: [{ type: i0.Input }],
            model: [{ type: i0.Input }],
            options: [{ type: i0.Input }],
            advancedFilters: [{ type: i0.Input }],
            isHistoryEnable: [{ type: i0.Input }],
            filterChange: [{ type: i0.Output }],
            onpopstate: [{ type: i0.HostListener, args: ['window:popstate', [''],] }]
        };
        return SdsFiltersComponent;
    }());

    var SdsAdvancedFiltersService = /** @class */ (function () {
        function SdsAdvancedFiltersService() {
        }
        SdsAdvancedFiltersService.prototype.convertToCheckboxes = function (origFields) {
            var _this = this;
            var fields = [];
            origFields.forEach(function (origField) {
                if (origField.fieldGroup && origField.fieldGroup.length > 1) {
                    var field = _this.createMulticheckbox(origField);
                    fields.push(field);
                }
                else {
                    var field = {
                        type: 'checkbox',
                        key: origField.key,
                        defaultValue: !origField.hide,
                        templateOptions: {
                            hideOptional: true
                        }
                    };
                    if (origField.templateOptions && origField.templateOptions.label) {
                        field.templateOptions.label = origField.templateOptions.label;
                    }
                    fields.push(field);
                }
            });
            return fields;
        };
        // TODO: Should be changed so option has label field instead of key but multicheckbox field type must be updated so default value still works
        SdsAdvancedFiltersService.prototype.createMulticheckbox = function (origField) {
            var options = [];
            var defaultValue = [];
            origField.fieldGroup.forEach(function (field) {
                var label = field.templateOptions && field.templateOptions.label ? field.templateOptions.label : null;
                var option = {
                    key: field.key,
                    value: label
                };
                options.push(option);
                if (!origField.hide && !field.hide) {
                    defaultValue.push(field.key);
                }
            });
            var field = {
                key: origField.key,
                type: 'multicheckbox',
                templateOptions: {
                    hideOptional: true,
                    type: 'array',
                    options: options
                }
            };
            if (origField.templateOptions && origField.templateOptions.label) {
                field.templateOptions.label = origField.templateOptions.label;
            }
            if (!origField.hide) {
                field.defaultValue = defaultValue;
            }
            return field;
        };
        SdsAdvancedFiltersService.prototype.updateFields = function (selectedFields, fields, model) {
            var _this = this;
            fields.forEach(function (field) {
                var key = field.key;
                var selectedField = selectedFields[key];
                if (field.fieldGroup && field.fieldGroup.length > 1) {
                    var fieldModel = model[key];
                    _this.updateFieldGroup(field, selectedField, fieldModel);
                }
                else {
                    _this.updateSingleField(field, selectedField, model);
                }
            });
            return {
                fields: fields,
                model: model
            };
        };
        SdsAdvancedFiltersService.prototype.updateFieldGroup = function (parentField, selectedFields, model) {
            var _this = this;
            if (selectedFields && selectedFields.length) {
                parentField.hide = false;
                parentField.fieldGroup.forEach(function (field) {
                    var key = field.key;
                    var fieldSelected = selectedFields.includes(key);
                    _this.updateSingleField(field, fieldSelected, model);
                });
            }
            else {
                parentField.hide = true;
                parentField.fieldGroup.forEach(function (field) {
                    _this.updateSingleField(field, false, model);
                });
            }
        };
        SdsAdvancedFiltersService.prototype.updateSingleField = function (field, fieldSelected, model) {
            if (fieldSelected) {
                field.hide = false;
            }
            else {
                field.hide = true;
                field.templateOptions['required'] = false;
                model[field.key] = null;
            }
        };
        SdsAdvancedFiltersService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        SdsAdvancedFiltersService.ctorParameters = function () { return []; };
        SdsAdvancedFiltersService.ngInjectableDef = i0.defineInjectable({ factory: function SdsAdvancedFiltersService_Factory() { return new SdsAdvancedFiltersService(); }, token: SdsAdvancedFiltersService, providedIn: "root" });
        return SdsAdvancedFiltersService;
    }());

    var SdsFormlyDialogComponent = /** @class */ (function () {
        function SdsFormlyDialogComponent(advancedFiltersService, dialogRef, data) {
            this.advancedFiltersService = advancedFiltersService;
            this.dialogRef = dialogRef;
            this.data = data;
        }
        SdsFormlyDialogComponent.prototype.ngOnInit = function () {
            this.fields = this.data.fields;
            this.form = this.data.form ? this.data.form : new forms.FormGroup({});
            this.model = this.data.model ? this.data.model : {};
            this.options = this.data.options ? this.data.options : {};
            this.cancel = this.data.cancel ? this.data.cancel : 'Cancel';
            this.submit = this.data.submit ? this.data.submit : 'Submit';
        };
        SdsFormlyDialogComponent.prototype.onSubmit = function () {
            if (this.form.valid) {
                var results = this.advancedFiltersService.updateFields(this.model, this.data.originalFields, this.data.originalModel);
                this.dialogRef.close(results);
            }
        };
        SdsFormlyDialogComponent.prototype.onCancel = function () {
            this.dialogRef.close();
        };
        SdsFormlyDialogComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-formly-dialog',
                        template: "<form [formGroup]=\"form\" (ngSubmit)=\"onSubmit()\">\n  <div *ngIf=\"data.title\" sds-dialog-title>{{data.title}}</div>\n  <div *ngIf=\"data.subtitle\" sds-dialog-subtitle>{{data.subtitle}}</div>\n  <div sds-dialog-content>\n    <formly-form [form]=\"form\" [fields]=\"fields\" [options]=\"options\" [model]=\"model\"></formly-form>\n  </div>\n  <div sds-dialog-actions>\n    <button class=\"usa-button usa-button--base\" (click)=\"onCancel()\" type=\"button\">\n      {{cancel}}\n    </button>\n    <button class=\"usa-button\" type=\"submit\">\n      {{submit}}\n    </button>\n  </div>\n</form>\n"
                    }] }
        ];
        /** @nocollapse */
        SdsFormlyDialogComponent.ctorParameters = function () {
            return [
                { type: SdsAdvancedFiltersService },
                { type: components.SdsDialogRef },
                { type: undefined, decorators: [{ type: i0.Inject, args: [components.SDS_DIALOG_DATA,] }] }
            ];
        };
        return SdsFormlyDialogComponent;
    }());

    var AdvancedFiltersComponent = /** @class */ (function () {
        function AdvancedFiltersComponent(dialog, advancedFiltersService) {
            this.dialog = dialog;
            this.advancedFiltersService = advancedFiltersService;
            /**
             *    Options for the form.
             */
            this.options = {};
        }
        AdvancedFiltersComponent.prototype.openDialog = function () {
            var _this = this;
            var modalFields = this.advancedFiltersService.convertToCheckboxes(this.fields);
            var data = {
                fields: modalFields,
                originalFields: this.fields,
                originalModel: this.model,
                submit: 'Update',
                title: 'More Filters',
            };
            var dialogRef = this.dialog.open(SdsFormlyDialogComponent, {
                width: 'medium',
                data: data
            });
            dialogRef.afterClosed().subscribe(function (result) {
                if (result) {
                    _this.fields = result.fields;
                    _this.model = result.model;
                }
            });
        };
        AdvancedFiltersComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'sds-advanced-filters',
                        template: "<button type=\"button\" (click)=\"openDialog()\" class=\"usa-button usa-button--unstyled\">More\n  Filters\n  <fa-icon [icon]=\"['sds', 'filter']\" [classes]=\"['icon-filter']\"></fa-icon>\n</button>\n",
                        styles: [".usa-button--unstyled{padding-left:0;width:initial;text-decoration:none;font-weight:600}.usa-button--unstyled:active,.usa-button--unstyled:hover{background:initial;color:initial;text-decoration:underline}"]
                    }] }
        ];
        /** @nocollapse */
        AdvancedFiltersComponent.ctorParameters = function () {
            return [
                { type: components.SdsDialogService },
                { type: SdsAdvancedFiltersService }
            ];
        };
        AdvancedFiltersComponent.propDecorators = {
            form: [{ type: i0.Input }],
            fields: [{ type: i0.Input }],
            model: [{ type: i0.Input }],
            options: [{ type: i0.Input }]
        };
        return AdvancedFiltersComponent;
    }());

    var SdsFiltersModule = /** @class */ (function () {
        function SdsFiltersModule() {
        }
        SdsFiltersModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [
                            SdsFiltersComponent,
                            AdvancedFiltersComponent,
                        ],
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            components.SdsAccordionModule,
                            SdsFormlyModule,
                            SdsFormlyResetModule,
                            angularFontawesome.FontAwesomeModule,
                            forms.ReactiveFormsModule,
                            select.FormlySelectModule,
                            core$1.FormlyModule
                        ],
                        exports: [SdsFiltersComponent]
                    },] }
        ];
        return SdsFiltersModule;
    }());

    var SdsFormlyDialogModule = /** @class */ (function () {
        function SdsFormlyDialogModule() {
        }
        SdsFormlyDialogModule.decorators = [
            { type: i0.NgModule, args: [{
                        declarations: [SdsFormlyDialogComponent],
                        imports: [
                            common.CommonModule,
                            SdsFiltersModule,
                            components.SdsDialogModule,
                            forms.ReactiveFormsModule,
                            SdsFormlyModule,
                            core$1.FormlyModule,
                            forms.FormsModule
                        ],
                        entryComponents: [SdsFormlyDialogComponent],
                        exports: [SdsFormlyDialogComponent]
                    },] }
        ];
        return SdsFormlyDialogModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.w = AdvancedFiltersComponent;
    exports.v = SdsFiltersComponent;
    exports.u = SdsFormlyResetComponent;
    exports.i = FormlyFieldAutoCompleteComponent;
    exports.l = FormlyFieldButtonComponent;
    exports.b = FormlyFieldCheckboxComponent;
    exports.j = FormlyFieldDatePickerComponent;
    exports.a = FormlyFieldInputComponent;
    exports.f = FormlyFieldMultiCheckboxComponent;
    exports.c = FormlyFieldRadioComponent;
    exports.s = FormlyFieldSearchComponent;
    exports.d = FormlyFieldSelectComponent;
    exports.q = FormlyFieldTextComponent;
    exports.e = FormlyFieldTextAreaComponent;
    exports.m = FormlyCustomWrapperComponent;
    exports.o = FormlyDescriptionWrapperComponent;
    exports.h = FormlyAccordianFormFieldComponent;
    exports.t = AnimationWrapperComponent;
    exports.k = FormlyFormFieldFilterWrapperComponent;
    exports.g = FormlyWrapperFormFieldComponent;
    exports.r = FormlyGroupWrapperComponent;
    exports.n = FormlyLabelWrapperComponent;
    exports.p = FormlyValidationWrapperComponent;
    exports.minlengthValidationMessage = minlengthValidationMessage;
    exports.maxlengthValidationMessage = maxlengthValidationMessage;
    exports.minValidationMessage = minValidationMessage;
    exports.minDateValidationMessage = minDateValidationMessage;
    exports.maxDateValidationMessage = maxDateValidationMessage;
    exports.betweenDateValidationMessage = betweenDateValidationMessage;
    exports.invalidDateFormatValidationMessage = invalidDateFormatValidationMessage;
    exports.maxValidationMessage = maxValidationMessage;
    exports.animationExtension = animationExtension;
    exports.SdsFormlyModule = SdsFormlyModule;
    exports.maxDateValidator = maxDateValidator;
    exports.minDateValidator = minDateValidator;
    exports.SdsFormlyResetModule = SdsFormlyResetModule;
    exports.SdsFiltersModule = SdsFiltersModule;
    exports.SdsFormlyDialogComponent = SdsFormlyDialogComponent;
    exports.SdsFormlyDialogModule = SdsFormlyDialogModule;
    exports.SDSFormlyUpdateComunicationService = SDSFormlyUpdateComunicationService;
    exports.SdsAdvancedFiltersService = SdsAdvancedFiltersService;
    exports.sdsFieldWrapper = sdsFieldWrapper;
    exports.sdsGroupWrapper = sdsGroupWrapper;
    exports.sdsWrappers = sdsWrappers;
    exports.FIELD_TYPE_COMPONENTS = FIELD_TYPE_COMPONENTS;
    exports.FORMLY_WRAPPERS = FORMLY_WRAPPERS;
    exports.FORMLY_CONFIG = FORMLY_CONFIG;
    exports.minDateToDateRangePicker = minDateToDateRangePicker;
    exports.minDateFromDateRangePicker = minDateFromDateRangePicker;
    exports.maxDateToDateRangePicker = maxDateToDateRangePicker;
    exports.maxDateFromDateRangePicker = maxDateFromDateRangePicker;
    exports.AbstractSdsFormly = AbstractSdsFormly;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
