// @gsa-sam/sam-formly/esm2015/lib/formly-dialog/formly-dialog.component.js
import { Component, Inject } from '@angular/core';
import { FormGroup } from '@angular/forms';
import { SdsDialogRef, SDS_DIALOG_DATA } from '@gsa-sam/components';
import { SdsAdvancedFiltersService } from '../formly-filters/advanced-filters/sds-advanced-filters.service';
export class SdsFormlyDialogComponent {
    constructor(advancedFiltersService, dialogRef, data) {
        this.advancedFiltersService = advancedFiltersService;
        this.dialogRef = dialogRef;
        this.data = data;
    }
    ngOnInit() {
        this.fields = this.data.fields;
        this.form = this.data.form ? this.data.form : new FormGroup({});
        this.model = this.data.model ? this.data.model : {};
        this.options = this.data.options ? this.data.options : {};
        this.cancel = this.data.cancel ? this.data.cancel : 'Cancel';
        this.submit = this.data.submit ? this.data.submit : 'Submit';
    }
    onSubmit() {
        if (this.form.valid) {
            const results = this.advancedFiltersService.updateFields(this.model, this.data.originalFields, this.data.originalModel);
            this.dialogRef.close(results);
        }
    }
    onCancel() {
        this.dialogRef.close();
    }
}
SdsFormlyDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-formly-dialog',
                template: "<form [formGroup]=\"form\" (ngSubmit)=\"onSubmit()\">\n  <div *ngIf=\"data.title\" sds-dialog-title>{{data.title}}</div>\n  <div *ngIf=\"data.subtitle\" sds-dialog-subtitle>{{data.subtitle}}</div>\n  <div sds-dialog-content>\n    <formly-form [form]=\"form\" [fields]=\"fields\" [options]=\"options\" [model]=\"model\"></formly-form>\n  </div>\n  <div sds-dialog-actions>\n    <button class=\"usa-button usa-button--base\" (click)=\"onCancel()\" type=\"button\">\n      {{cancel}}\n    </button>\n    <button class=\"usa-button\" type=\"submit\">\n      {{submit}}\n    </button>\n  </div>\n</form>\n"
            }] }
];
/** @nocollapse */
SdsFormlyDialogComponent.ctorParameters = () => [
    { type: SdsAdvancedFiltersService },
    { type: SdsDialogRef },
    { type: undefined, decorators: [{ type: Inject, args: [SDS_DIALOG_DATA,] }] }
];
