// @gsa-sam/components/esm2015/lib/popup/popup.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, ElementRef, Renderer2 } from '@angular/core';
export class SdsPopupDirective {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.renderer.addClass(this.el.nativeElement, 'sds-popup');
        this.sdsPopupDiv = document.createElement('div');
        this.renderer.addClass(this.sdsPopupDiv, 'sds-popup__content');
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.renderer.appendChild(this.el.nativeElement, this.sdsPopup);
        this.renderer.addClass(this.sdsPopupDiv, this.placement);
        this.renderer.addClass(this.sdsPopupDiv, this.position);
        this.renderer.appendChild(this.sdsPopupDiv, this.el.nativeElement.children[0]);
        this.renderer.appendChild(this.el.nativeElement, this.sdsPopupDiv);
    }
}
SdsPopupDirective.decorators = [
    { type: Directive, args: [{
                selector: '[sdsPopup]',
                exportAs: 'sdsPopup'
            },] }
];
/** @nocollapse */
SdsPopupDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
SdsPopupDirective.propDecorators = {
    sdsPopup: [{ type: Input }],
    position: [{ type: Input }],
    placement: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SdsPopupDirective.prototype.sdsPopup;
    /** @type {?} */
    SdsPopupDirective.prototype.position;
    /** @type {?} */
    SdsPopupDirective.prototype.placement;
    /** @type {?} */
    SdsPopupDirective.prototype.sdsPopupDiv;
    /**
     * @type {?}
     * @private
     */
    SdsPopupDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    SdsPopupDirective.prototype.renderer;
}
