// @gsa-sam/components/esm2015/lib/toolbar/toolbar-expand.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2 } from "@angular/core";
import { SdsToolbarComponent } from "./toolbar.component";
export class SdsToolbarExpandDirective {
    /**
     * @param {?} renderer
     * @param {?} _element
     */
    constructor(renderer, _element) {
        this.renderer = renderer;
        this._element = _element;
    }
    /**
     * References the toolbar instance that the element its associated with.
     * @return {?}
     */
    get toolbar() {
        return this._toolbar;
    }
    /**
     * @param {?} toolbar
     * @return {?}
     */
    set toolbar(toolbar) {
        if (toolbar === this._toolbar) {
            return;
        }
        this._toolbar = toolbar;
        /** Sets toolbar expanded width */
        this._expandedWidth = this._toolbar.expandedSpace;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setStyle(this.toolbar.expanded);
        this.toolbar.expandedChange.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => {
            this.setStyle(value);
        }));
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    setStyle(value) {
        if (value) {
            this.renderer.setStyle(this._element.nativeElement, "margin-left", `${this._expandedWidth}`);
        }
        else {
            this.renderer.removeStyle(this._element.nativeElement, "margin-left");
        }
    }
}
SdsToolbarExpandDirective.decorators = [
    { type: Directive, args: [{
                selector: "[sdsToolbarExpand]"
            },] }
];
/** @nocollapse */
SdsToolbarExpandDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
SdsToolbarExpandDirective.propDecorators = {
    toolbar: [{ type: Input, args: ["sdsToolbarExpand",] }]
};
if (false) {
    /**
     * Width of the toolbar while expanded.
     * @type {?}
     */
    SdsToolbarExpandDirective.prototype._expandedWidth;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarExpandDirective.prototype._toolbar;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarExpandDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarExpandDirective.prototype._element;
}
