// @gsa-sam/components/esm2015/lib/toolbar/toolbar-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Host } from "@angular/core";
import { SdsToolbarComponent } from "./toolbar.component";
export class SdsToolbarHeaderComponent {
    /**
     * @param {?} toolbar
     */
    constructor(toolbar) {
        this.toolbar = toolbar;
    }
    /**
     * Toggles the expanded state of the toolbar.
     * @return {?}
     */
    _toggle() {
        this.toolbar.toggle();
    }
    /**
     * Gets whether the toolbar is expanded.
     * @return {?}
     */
    get _isExpanded() {
        return this.toolbar.expanded;
    }
    /**
     * Whether the toolbar is disabled.
     * @return {?}
     */
    get _disabled() {
        return this.toolbar.disabled;
    }
}
SdsToolbarHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: "sds-toolbar-header",
                template: "<button\n  class=\"sds-toolbar__toogle-btn\"\n  [attr.id]=\"toolbar._headerId\"\n  [attr.tabindex]=\"_disabled ? -1 : 0\"\n  [attr.aria-controls]=\"toolbar._contentId\"\n  [attr.aria-expanded]=\"_isExpanded\"\n  [attr.aria-disabled]=\"_disabled\"\n  (click)=\"_toggle()\"\n>\n  <div class=\"sds-toolbar__toggle\">\n    <span class=\"usa-sr-only\">Open/Close Toolbar</span>\n  </div>\n</button>\n",
                host: {
                    class: "sds-toolbar__header"
                }
            }] }
];
/** @nocollapse */
SdsToolbarHeaderComponent.ctorParameters = () => [
    { type: SdsToolbarComponent, decorators: [{ type: Host }] }
];
if (false) {
    /** @type {?} */
    SdsToolbarHeaderComponent.prototype.toolbar;
}
