// @gsa-sam/components/esm2015/lib/toolbar/toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { coerceBooleanProperty } from "@angular/cdk/coercion";
import { sdsToolbarAnimations } from "./toolbar-animations";
/**
 * Counter for generating unique element ids.
 * @type {?}
 */
let uniqueId = 0;
export class SdsToolbarComponent {
    constructor() {
        /**
         * The unique Toolbar id.
         */
        this.id = uniqueId++;
        /**
         * ID for the header element. Used for a11y labelling.
         */
        this._headerId = `sds-toolbar-header-${this.id}`;
        /**
         * ID for the content element. Used for a11y labelling.
         */
        this._contentId = `sds-toolbar-content-${this.id}`;
        /**
         * Emits whenever the expanded state of the toolbar changes.
         */
        this.expandedChange = new EventEmitter();
        this._expanded = false;
        this._disabled = false;
        /**
         * Width of the toolbar when is expanded.
         */
        this.expandedWidth = "300px";
        // Elements that surround the toolbar need to make space for the expanded toolbar.
        // Adding 20 more pixes to provide padding
        this.expandedSpace = "20px";
    }
    /**
     * Whether the toolbar is expanded.
     * @return {?}
     */
    get expanded() {
        return this._expanded;
    }
    /**
     * @param {?} expanded
     * @return {?}
     */
    set expanded(expanded) {
        expanded = coerceBooleanProperty(expanded);
        // Only emit events and update the internal value if the value changes.
        if (this._expanded !== expanded) {
            this._expanded = expanded;
            this.expandedChange.emit(expanded);
        }
    }
    /**
     * Whether the toolbar is disabled.
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
    /**
     * @param {?} disabled
     * @return {?}
     */
    set disabled(disabled) {
        this._disabled = coerceBooleanProperty(disabled);
    }
    /**
     * Toggles the expanded state of the toolbar.
     * @return {?}
     */
    toggle() {
        this.expanded = !this.expanded;
    }
    /**
     * Gets the expanded state string.
     * @return {?}
     */
    _getExpandedState() {
        return this.expanded ? "expanded" : "collapsed";
    }
}
SdsToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: "sds-toolbar",
                exportAs: "sdsToolbar",
                template: "<sds-toolbar-header></sds-toolbar-header>\n<div\n  class=\"sds-toolbar__content\"\n  role=\"region\"\n  [@bodyExpansion]=\"{\n    value: _getExpandedState(),\n    params: { expandedWidth: expandedWidth }\n  }\"\n  [attr.aria-labelledby]=\"_headerId\"\n  [id]=\"_contentId\"\n>\n  <ng-content></ng-content>\n</div>\n",
                animations: [sdsToolbarAnimations.bodyExpansion],
                host: {
                    class: "sds-toolbar",
                    "[class.sds-toolbar--expanded]": "expanded"
                }
            }] }
];
/** @nocollapse */
SdsToolbarComponent.ctorParameters = () => [];
SdsToolbarComponent.propDecorators = {
    expandedChange: [{ type: Output }],
    expanded: [{ type: Input }],
    disabled: [{ type: Input }],
    expandedWidth: [{ type: Input }],
    expandedSpace: [{ type: Input }]
};
if (false) {
    /**
     * The unique Toolbar id.
     * @type {?}
     */
    SdsToolbarComponent.prototype.id;
    /**
     * ID for the header element. Used for a11y labelling.
     * @type {?}
     */
    SdsToolbarComponent.prototype._headerId;
    /**
     * ID for the content element. Used for a11y labelling.
     * @type {?}
     */
    SdsToolbarComponent.prototype._contentId;
    /**
     * Emits whenever the expanded state of the toolbar changes.
     * @type {?}
     */
    SdsToolbarComponent.prototype.expandedChange;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarComponent.prototype._expanded;
    /**
     * @type {?}
     * @private
     */
    SdsToolbarComponent.prototype._disabled;
    /**
     * Width of the toolbar when is expanded.
     * @type {?}
     */
    SdsToolbarComponent.prototype.expandedWidth;
    /** @type {?} */
    SdsToolbarComponent.prototype.expandedSpace;
}
