// @gsa-sam/sam-formly/esm5/lib/formly-reset/formly-reset.component.js
import { Component, Input } from '@angular/core';
var SdsFormlyResetComponent = /** @class */ (function () {
    function SdsFormlyResetComponent() {
        /**
         * Pass in classes for reset button -- default .usa-button .usa-button--unstyled
         */
        this.classes = ['usa-button', 'usa-button--unstyled'];
    }
    SdsFormlyResetComponent.prototype.resetAll = function () {
        this.options.resetModel();
    };
    SdsFormlyResetComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-formly-reset',
                    template: "<button (click)=\"resetAll()\" [ngClass]=\"classes\">\n  Reset All\n  <fa-icon [icon]=\"['sds', 'reset']\" [classes]=\"['icon-reset']\"></fa-icon>\n</button>\n",
                    styles: [".usa-button--unstyled{padding-right:0;width:initial;text-decoration:none;font-weight:600}.usa-button--unstyled:active,.usa-button--unstyled:hover{background:initial;color:initial;text-decoration:underline}"]
                }] }
    ];
    SdsFormlyResetComponent.propDecorators = {
        options: [{ type: Input }],
        classes: [{ type: Input }]
    };
    return SdsFormlyResetComponent;
}());
export { SdsFormlyResetComponent };
