// @gsa-sam/components/esm2015/lib/truncate-text/truncate-text-container.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, HostBinding, HostListener } from '@angular/core';
import { SDS_TRUNCATED_TEXT_DATA } from './truncates-text-base';
import { sdsTruncateTextAnimations } from './truncate-text-animations';
/**
 * @record
 */
export function SdsTruncateTextData() { }
if (false) {
    /** @type {?} */
    SdsTruncateTextData.prototype.text;
}
export class SdsTruncatedTextContainerComponent {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = data;
        this._animationState = 'void';
    }
    /**
     * Starts the animation.
     * @return {?}
     */
    startAnimation() {
        this._animationState = 'enter';
    }
    /**
     * Resets the animation to its initial state.
     * @return {?}
     */
    resetAnimation() {
        this._animationState = 'void';
    }
    /**
     * Intentionally left empty to trigger change detection
     * @return {?}
     */
    _onAnimationDone() { }
}
SdsTruncatedTextContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-truncated-text-container',
                template: `
    <div class="sds-overlay maxw-mobile radius-overlay padding-2">{{ data.text }}</div>
  `,
                animations: [sdsTruncateTextAnimations.container]
            }] }
];
/** @nocollapse */
SdsTruncatedTextContainerComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [SDS_TRUNCATED_TEXT_DATA,] }] }
];
SdsTruncatedTextContainerComponent.propDecorators = {
    _animationState: [{ type: HostBinding, args: ['@container',] }],
    _onAnimationDone: [{ type: HostListener, args: ['@container.done',] }]
};
if (false) {
    /** @type {?} */
    SdsTruncatedTextContainerComponent.prototype._animationState;
    /** @type {?} */
    SdsTruncatedTextContainerComponent.prototype.data;
}
