// @gsa-sam/sam-formly/esm5/lib/formly/types/checkbox.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
var FormlyFieldCheckboxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FormlyFieldCheckboxComponent, _super);
    function FormlyFieldCheckboxComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultOptions = {
            templateOptions: {
                indeterminate: true,
                hideLabel: true,
            },
        };
        return _this;
    }
    FormlyFieldCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-formly-field-checkbox',
                    template: "\n    <div class=\"usa-checkbox\">\n      <input class=\"usa-checkbox__input\" type=\"checkbox\"\n        [class.is-invalid]=\"showError\"\n        [indeterminate]=\"to.indeterminate && formControl.value === null\"\n        [formControl]=\"formControl\"\n        [formlyAttributes]=\"field\">\n      <label class=\"usa-checkbox__label\" [for]=\"id\">\n        {{ to.label }}\n        <span *ngIf=\"!to.required && !to.hideOptional\"> (Optional)</span>\n      </label>\n    </div>\n  "
                }] }
    ];
    return FormlyFieldCheckboxComponent;
}(FieldType));
export { FormlyFieldCheckboxComponent };
