// @gsa-sam/sam-formly/esm5/lib/formly/types/multicheckbox.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
var FormlyFieldMultiCheckboxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FormlyFieldMultiCheckboxComponent, _super);
    function FormlyFieldMultiCheckboxComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultOptions = {
            templateOptions: {
                options: [],
            },
        };
        return _this;
    }
    FormlyFieldMultiCheckboxComponent.prototype.onChange = function (value, checked) {
        var _a;
        if (this.to.type === 'array') {
            this.formControl.patchValue(checked
                ? tslib_1.__spread((this.formControl.value || []), [value]) : tslib_1.__spread((this.formControl.value || [])).filter(function (o) { return o !== value; }));
        }
        else {
            this.formControl.patchValue(tslib_1.__assign({}, this.formControl.value, (_a = {}, _a[value] = checked, _a)));
        }
        this.formControl.markAsTouched();
    };
    FormlyFieldMultiCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-formly-field-multicheckbox',
                    template: "\n  <div class=\"usa-checkbox\">\n    <div *ngFor=\"let option of to.options | formlySelectOptions:field | async; let i = index;\">\n    <input type=\"checkbox\"\n     [id]=\"id + '_' + i\"\n     class=\"usa-checkbox__input\"\n      [value]=\"option.value\"\n      [checked]=\"formControl.value && (this.to.type === 'array' ? formControl.value.includes(option.value) : formControl.value[option.value])\"\n      [formlyAttributes]=\"field\"\n      (change)=\"onChange(option.value, $event.target.checked)\">\n    <label\n    class=\"usa-checkbox__label\"\n    [for]=\"id + '_' + i\">\n        {{ option.label }}\n    </label>\n  </div>\n</div>\n  "
                }] }
    ];
    return FormlyFieldMultiCheckboxComponent;
}(FieldType));
export { FormlyFieldMultiCheckboxComponent };
