// @gsa-sam/sam-formly/esm5/lib/formly/types/textarea.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
var FormlyFieldTextAreaComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FormlyFieldTextAreaComponent, _super);
    function FormlyFieldTextAreaComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultOptions = {
            templateOptions: {
                cols: 1,
                rows: 1,
            },
        };
        return _this;
    }
    FormlyFieldTextAreaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-formly-field-textarea',
                    template: "\n    <textarea [formControl]=\"formControl\" [cols]=\"to.cols\"\n      [rows]=\"to.rows\" class=\"usa-textarea\" [class.usa-input--error]=\"showError\"\n      [formlyAttributes]=\"field\">\n    </textarea>\n  "
                }] }
    ];
    return FormlyFieldTextAreaComponent;
}(FieldType));
export { FormlyFieldTextAreaComponent };
