// @gsa-sam/layouts/esm5/lib/feature/search-list-layout/search-list-layout.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChild, TemplateRef, Optional } from '@angular/core';
import { BehaviorSubject } from "rxjs";
import { SearchListConfiguration } from './model/search-list-layout.model';
import { SDSFormlyUpdateComunicationService } from '@gsa-sam/sam-formly';
var SearchListLayoutComponent = /** @class */ (function () {
    function SearchListLayoutComponent(formlyUpdateComunicationService) {
        this.formlyUpdateComunicationService = formlyUpdateComunicationService;
        /**
         * Default Page setttings
         */
        this.page = {
            pageNumber: 1,
            pageSize: 25,
            totalPages: 0
        };
        /**
         * Id of the top pagination control
         */
        this.top = { id: 'topPagination' };
        /**
         * Id of the bottom pagination control
         */
        this.bottom = { id: 'bottomPagination' };
        /**
         * Page event listener
         */
        this.paginationChange = new BehaviorSubject(this.page);
        /**
         * List of items to be displayed
         */
        this.items = [];
        /**
         * sort value
         */
        this.sortField = '';
    }
    /**
     * @return {?}
     */
    SearchListLayoutComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.page.pageSize = this.configuration.pageSize;
        this.sortField = this.configuration.defaultSortValue;
        this.paginationChange.subscribe((/**
         * @return {?}
         */
        function () {
            _this.updateContent();
        }));
        if (this.formlyUpdateComunicationService) {
            this.formlyUpdateComunicationService.filterUpdate.subscribe((/**
             * @param {?} filter
             * @return {?}
             */
            function (filter) {
                _this.updateFilter(filter);
            }));
        }
    };
    /**
     * updates the filter and set the page number to 1 and calls imported service
     * @param filter
     */
    /**
     * updates the filter and set the page number to 1 and calls imported service
     * @param {?} filter
     * @return {?}
     */
    SearchListLayoutComponent.prototype.updateFilter = /**
     * updates the filter and set the page number to 1 and calls imported service
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        this.filterData = filter;
        this.page.pageNumber = 1;
        this.updateContent();
    };
    /**
     * Sorty by change event
     */
    /**
     * Sorty by change event
     * @return {?}
     */
    SearchListLayoutComponent.prototype.onSelectChange = /**
     * Sorty by change event
     * @return {?}
     */
    function () {
        this.page.pageNumber = 1;
        this.updateContent();
    };
    /**
     * calls service when updated
     */
    /**
     * calls service when updated
     * @private
     * @return {?}
     */
    SearchListLayoutComponent.prototype.updateContent = /**
     * calls service when updated
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.service.getData({ 'page': this.page, sortField: this.sortField, filter: this.filterData }).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) {
            _this.items = result.items;
            _this.page.totalPages = Math.ceil(result.totalItems / _this.page.pageSize);
        }));
    };
    SearchListLayoutComponent.decorators = [
        { type: Component, args: [{
                    selector: 'search-list-layout',
                    template: "<div class=\"grid-row\" *ngIf=\"items?.length\">\n  <div class=\"grid-col\">\n    <div class=\"grid-row\">\n      <div class=\"grid-col-auto\">\n        <sds-pagination class=\"tablet:grid-col\" [paginationConfiguration]=\"top\" [(page)]=\"page\" (pageChange)=\"paginationChange.next($event)\">\n        </sds-pagination>\n      </div>\n      <div class=\"grid-col-fill\"></div>\n      <div class=\"grid-col-auto float-right\">\n        <div class=\"margin-left-1\">\n          <label class=\"usa-label font-sans-3xs text-italic text-base margin-top-1\" for=\"options\">\n            Sort by\n          </label>\n          <select (change)=\"onSelectChange()\" [(ngModel)]=\"sortField\" class=\"usa-select usa-select--small sds-min-width-160 border-base-light\" id=\"options\">\n            <option *ngFor=\"let item of configuration.sortList\" [ngValue]=\"item.value\">\n              {{ item.text }}\n            </option>\n          </select>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n<div class=\"grid-row\">\n  <div class=\"grid-col-12\">\n    <sds-search-result-list [model]=\"items\">\n      <ng-template #resultContent let-item>\n        <ng-container *ngTemplateOutlet=\"resultContentTemplate,\n                context: { $implicit: item }\"></ng-container>\n      </ng-template>\n    </sds-search-result-list>\n    <sds-pagination *ngIf=\"items?.length\" [paginationConfiguration]=\"bottom\" [(page)]=\"page\" (pageChange)=\"paginationChange.next($event)\">\n    </sds-pagination>\n  </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SearchListLayoutComponent.ctorParameters = function () { return [
        { type: SDSFormlyUpdateComunicationService, decorators: [{ type: Optional }] }
    ]; };
    SearchListLayoutComponent.propDecorators = {
        resultContentTemplate: [{ type: ContentChild, args: ['resultContent',] }],
        service: [{ type: Input }],
        configuration: [{ type: Input }]
    };
    return SearchListLayoutComponent;
}());
export { SearchListLayoutComponent };
if (false) {
    /**
     * Child Template to be used to display the data for each item in the list of items
     * @type {?}
     */
    SearchListLayoutComponent.prototype.resultContentTemplate;
    /**
     * Input service to be called when items change
     * @type {?}
     */
    SearchListLayoutComponent.prototype.service;
    /**
     * configuration
     * @type {?}
     */
    SearchListLayoutComponent.prototype.configuration;
    /**
     * Filter information
     * @type {?}
     * @private
     */
    SearchListLayoutComponent.prototype.filterData;
    /**
     * Default Page setttings
     * @type {?}
     */
    SearchListLayoutComponent.prototype.page;
    /**
     * Id of the top pagination control
     * @type {?}
     */
    SearchListLayoutComponent.prototype.top;
    /**
     * Id of the bottom pagination control
     * @type {?}
     */
    SearchListLayoutComponent.prototype.bottom;
    /**
     * Page event listener
     * @type {?}
     */
    SearchListLayoutComponent.prototype.paginationChange;
    /**
     * List of items to be displayed
     * @type {?}
     */
    SearchListLayoutComponent.prototype.items;
    /**
     * sort value
     * @type {?}
     */
    SearchListLayoutComponent.prototype.sortField;
    /**
     * @type {?}
     * @private
     */
    SearchListLayoutComponent.prototype.formlyUpdateComunicationService;
}
