// @gsa-sam/layouts/esm2015/lib/feature/subheader/drawer-communication.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
export class SdsDrawerCommunicationService {
    constructor() {
        this.isOpen = new BehaviorSubject(false);
        this.isDrawerOpen = this.isOpen.asObservable();
        this.currentTemplate = new BehaviorSubject(null);
        this.contentTemplate = this.currentTemplate.asObservable();
    }
    /**
     * @param {?} open
     * @param {?} temp
     * @return {?}
     */
    onDrawerOpen(open, temp) {
        this.isOpen.next(open);
        this.currentTemplate.next(temp);
    }
}
SdsDrawerCommunicationService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    SdsDrawerCommunicationService.prototype.isOpen;
    /** @type {?} */
    SdsDrawerCommunicationService.prototype.isDrawerOpen;
    /**
     * @type {?}
     * @private
     */
    SdsDrawerCommunicationService.prototype.currentTemplate;
    /** @type {?} */
    SdsDrawerCommunicationService.prototype.contentTemplate;
}
