// @gsa-sam/layouts/esm2015/lib/feature/subheader/subheader.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, Input, TemplateRef } from '@angular/core';
import { SdsDrawerCommunicationService } from './drawer-communication.service';
export class SdsSubheaderComponent {
    constructor() { }
}
SdsSubheaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-subheader',
                template: "<div class=\"sds-navbar border-bottom border-base-lighter\">\n  <div class=\"sds-navbar__content\">\n    <div class=\"grid-row\">\n      <div\n        class=\"grid-col-12 tablet-lg:grid-col-auto display-flex flex-align-center margin-bottom-1 tablet-lg:margin-bottom-0\"\n      >\n        <ng-content></ng-content>\n      </div>\n      <div class=\"grid-col-12 tablet-lg:grid-col-fill display-flex\">\n        <div class=\"tablet-lg:order-1\">\n          <!-- Buttons -->\n          <ng-content select=\"[subheader-buttons-container]\"></ng-content>\n        </div>\n        <div class=\"flex-fill\">\n          <!-- Search -->\n          <div class=\"margin-x-105 margin-top-05 tablet:margin-top-0\">\n            <ng-content select=\"sds-search\"></ng-content>\n          </div>\n        </div>\n        <div class=\"tablet-lg:order-2 margin-left-2\">\n          <!-- Actions Menu Trigger -->\n          <ng-content select=\"sds-subheader-actions\"></ng-content>\n\n          <!-- Drawer Trigger -->\n          <ng-content select=\"sds-subheader-drawer\"></ng-content>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n\n<!-- =============== DRAWER CONTENT =============== -->\n<sds-drawer-content></sds-drawer-content>\n"
            }] }
];
/** @nocollapse */
SdsSubheaderComponent.ctorParameters = () => [];
export class SdsSubheaderActionsComponent {
    constructor() {
        this.clicks = new EventEmitter();
    }
}
SdsSubheaderActionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-subheader-actions',
                template: "<!-- Actions Menu Trigger: Desktop -->\n<span class=\"display-none desktop-lg:display-inline-block\">\n  <button\n    class=\"sds-button sds-button--labeled-icon\"\n    [sdsMenuTriggerFor]=\"menu\"\n  >\n    Actions\n    <fa-layers [fixedWidth]=\"true\" size=\"lg\">\n      <fa-icon [icon]=\"['fas', 'circle']\"></fa-icon>\n      <fa-icon [icon]=\"['fas', 'ellipsis-v']\" transform=\"shrink-6\"></fa-icon>\n    </fa-layers>\n  </button>\n</span>\n\n<!-- Actions Menu Trigger: Tablet -->\n<span class=\"desktop-lg:display-none \">\n  <span class=\"display-none tablet-lg:display-inline-block\">\n    <button\n      class=\"sds-button sds-button--primary sds-button--circular\"\n      [sdsMenuTriggerFor]=\"menu\"\n    >\n      <fa-icon [fixedWidth]=\"true\" [icon]=\"['fas', 'ellipsis-v']\"></fa-icon>\n      <span class=\"usa-sr-only\">Toggle Actions Menu</span>\n    </button>\n  </span>\n</span>\n\n<!-- Actions Menu Trigger: Mobile -->\n<span class=\"tablet-lg:display-none\">\n  <button\n    *ngFor=\"let button of model.actions\"\n    (click)=\"clicks.emit(button.id)\"\n    class=\"sds-button sds-button--circular margin-top-05 tablet:margin-top-0\"\n  >\n    <fa-icon [fixedWidth]=\"true\" [icon]=\"['sds', button.icon]\"></fa-icon>\n  </button>\n</span>\n\n<!-- Menu content -->\n<sds-menu #menu=\"sdsMenu\" xPosition=\"before\" overlapTrigger=\"true\">\n  <!-- Menu header (optional) -->\n  <sds-menu-header>Actions</sds-menu-header>\n  <!-- Menu items -->\n  <button\n    *ngFor=\"let button of model.actions\"\n    (click)=\"clicks.emit(button.id)\"\n    sds-menu-item\n  >\n    {{ button.text }}\n  </button>\n</sds-menu>\n"
            }] }
];
/** @nocollapse */
SdsSubheaderActionsComponent.ctorParameters = () => [];
SdsSubheaderActionsComponent.propDecorators = {
    model: [{ type: Input }],
    clicks: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    SdsSubheaderActionsComponent.prototype.model;
    /** @type {?} */
    SdsSubheaderActionsComponent.prototype.clicks;
}
export class SdsSubheaderDrawerComponent {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = data;
        this.isDrawerOpen = new EventEmitter();
        this.isOpen = false;
    }
    /**
     * @param {?} ev
     * @return {?}
     */
    onDrawerOpenClose(ev) {
        this.isOpen = !this.isOpen;
        this.data.onDrawerOpen(this.isOpen, this.drawerContentTemplate);
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
SdsSubheaderDrawerComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-subheader-drawer',
                template: "<button\n  (click)=\"onDrawerOpenClose($event)\"\n  class=\"sds-button sds-button--circular margin-top-2px tablet:margin-top-0\"\n>\n  <fa-icon\n    [fixedWidth]=\"true\"\n    [icon]=\"isOpen ? ['fas', 'arrow-up'] : ['fas', 'arrow-down']\"\n  ></fa-icon>\n</button>\n"
            }] }
];
/** @nocollapse */
SdsSubheaderDrawerComponent.ctorParameters = () => [
    { type: SdsDrawerCommunicationService }
];
SdsSubheaderDrawerComponent.propDecorators = {
    drawerContentTemplate: [{ type: Input }],
    isDrawerOpen: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    SdsSubheaderDrawerComponent.prototype.drawerContentTemplate;
    /** @type {?} */
    SdsSubheaderDrawerComponent.prototype.isDrawerOpen;
    /** @type {?} */
    SdsSubheaderDrawerComponent.prototype.isOpen;
    /** @type {?} */
    SdsSubheaderDrawerComponent.prototype.data;
}
export class SdsDrawerContentComponent {
    /**
     * @param {?} data
     */
    constructor(data) {
        this.data = data;
        this.isDrawerOpen = false;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.data.contentTemplate.subscribe((/**
         * @param {?} template
         * @return {?}
         */
        template => (this.drawerContentTemplate = template)));
        this.data.isDrawerOpen.subscribe((/**
         * @param {?} open
         * @return {?}
         */
        open => (this.isDrawerOpen = open)));
    }
}
SdsDrawerContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-drawer-content',
                template: "<ng-container *ngIf=\"drawerContentTemplate && isDrawerOpen\" [ngTemplateOutlet]=\"drawerContentTemplate\" >\n</ng-container>\n"
            }] }
];
/** @nocollapse */
SdsDrawerContentComponent.ctorParameters = () => [
    { type: SdsDrawerCommunicationService }
];
if (false) {
    /** @type {?} */
    SdsDrawerContentComponent.prototype.drawerContentTemplate;
    /** @type {?} */
    SdsDrawerContentComponent.prototype.isDrawerOpen;
    /** @type {?} */
    SdsDrawerContentComponent.prototype.data;
}
