// @gsa-sam/components/esm2015/lib/selected-result/models/sds-selected-item-model-helper.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class SDSSelectedItemModelHelper {
    /**
     *  adds an item to the collection
     * if selected mode is single it removes any existing items
     * also checks to see if that item already exists
     * keyfield is used to determine uniqueness of the item added
     * @param {?} itemToAdd
     * @param {?} keyField
     * @param {?} selectionMode
     * @param {?} items
     * @return {?}
     */
    static addItem(itemToAdd, keyField, selectionMode, items) {
        if (!SDSSelectedItemModelHelper.contatinsItem(itemToAdd[keyField], keyField, items)) {
            if (selectionMode === SelectionMode.SINGLE) {
                SDSSelectedItemModelHelper.clearItems(items);
            }
            items.push(itemToAdd);
        }
    }
    /**
     * adds an array of items to the list and will not add duplicate items
     * keyfield is used to determine uniqueness of the item added
     * @param {?} toAddItems
     * @param {?} keyField
     * @param {?} selectionMode
     * @param {?} items
     * @return {?}
     */
    static addItems(toAddItems, keyField, selectionMode, items) {
        for (let i = 0; i < toAddItems.length; i++) {
            SDSSelectedItemModelHelper.addItem(toAddItems[i], keyField, selectionMode, items);
        }
    }
    /**
     * removes the item from the list
     * keyfield is used to determine uniqueness of the item added
     * @param {?} item
     * @param {?} keyField
     * @param {?} items
     * @return {?}
     */
    static removeItem(item, keyField, items) {
        if (SDSSelectedItemModelHelper.contatinsItem(item[keyField], keyField, items)) {
            items.splice(items.indexOf(item), 1);
        }
    }
    /**
     * checks to see if a particular item exists by the provied key
     * keyfield is used to determine uniqueness of the item added
     * @param {?} key
     * @param {?} keyField
     * @param {?} items
     * @return {?}
     */
    static contatinsItem(key, keyField, items) {
        /** @type {?} */
        let item = items.find((/**
         * @param {?} o
         * @return {?}
         */
        o => o[keyField] === key));
        return item !== null && item !== undefined;
    }
    /**
     * Clears the list of items
     * @param {?} items
     * @return {?}
     */
    static clearItems(items) {
        while (items.length > 0) {
            items.pop();
        }
    }
    /**
     * updates an array of items to the list and will not add duplicate items
     * keyfield is used to determine uniqueness of the item added
     * @param {?} selectedItems
     * @param {?} keyField
     * @param {?} selectionMode
     * @param {?} items
     * @return {?}
     */
    static replaceItems(selectedItems, keyField, selectionMode, items) {
        //Clears Old List
        SDSSelectedItemModelHelper.clearItems(items);
        //Adds new List
        SDSSelectedItemModelHelper.addItems(selectedItems, keyField, selectionMode, items);
    }
}
/** @enum {number} */
const SelectionMode = {
    SINGLE: 0, MULTIPLE: 1,
};
export { SelectionMode };
SelectionMode[SelectionMode.SINGLE] = 'SINGLE';
SelectionMode[SelectionMode.MULTIPLE] = 'MULTIPLE';
