// @gsa-sam/layouts/esm5/lib/feature/actions-menu/actions-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
var SdsActionsMenuComponent = /** @class */ (function () {
    function SdsActionsMenuComponent() {
        this.clicks = new EventEmitter();
    }
    SdsActionsMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-actions-menu',
                    template: "<!-- Button that triggers menu (sdsMenuTriggerFor) -->\n<button\n  class=\"sds-button sds-button--circular\"\n  [class.sds-button--primary]=\"model.trigger.type === 'primary'\"\n  [class.sds-button--shadow]=\"model.trigger.shadow\"\n  [class.sds-button--small]=\"size === 'sm'\"\n  [sdsMenuTriggerFor]=\"menu\"\n>\n  <fa-icon\n    [fixedWidth]=\"true\"\n    [icon]=\"['sds', 'ellipsis']\"\n    transform=\"grow-5\"\n  ></fa-icon>\n  <span class=\"usa-sr-only\">Toggle Actions Menu</span>\n</button>\n\n<!-- Menu content -->\n<sds-menu\n  #menu=\"sdsMenu\"\n  [size]=\"size\"\n  xPosition=\"before\"\n  overlapTrigger=\"true\"\n>\n  <!-- Menu header (optional) -->\n  <sds-menu-header>Actions</sds-menu-header>\n  <!-- Menu items -->\n  <button\n    *ngFor=\"let button of model.actions\"\n    (click)=\"clicks.emit(button.id)\"\n    sds-menu-item\n  >\n    {{ button.text }}\n  </button>\n</sds-menu>\n"
                }] }
    ];
    /** @nocollapse */
    SdsActionsMenuComponent.ctorParameters = function () { return []; };
    SdsActionsMenuComponent.propDecorators = {
        model: [{ type: Input }],
        size: [{ type: Input }],
        clicks: [{ type: Output }]
    };
    return SdsActionsMenuComponent;
}());
export { SdsActionsMenuComponent };
if (false) {
    /** @type {?} */
    SdsActionsMenuComponent.prototype.model;
    /** @type {?} */
    SdsActionsMenuComponent.prototype.size;
    /** @type {?} */
    SdsActionsMenuComponent.prototype.clicks;
}
