// @gsa-sam/components/esm5/lib/external-link/external-link.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Renderer2, Input, ComponentFactoryResolver, ViewContainerRef } from '@angular/core';
import { FaIconComponent } from '@fortawesome/angular-fontawesome';
var ExternalLinkDirective = /** @class */ (function () {
    function ExternalLinkDirective(el, renderer, cfr, vc) {
        this.el = el;
        this.renderer = renderer;
        this.cfr = cfr;
        this.vc = vc;
        this.hideIcon = false;
    }
    Object.defineProperty(ExternalLinkDirective.prototype, "isExternalLink", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            return this.href.replace(/^https?:\/\//, '').replace(/^www\./, '').split('/')[0] != location.hostname;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ExternalLinkDirective.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (!this.isExternalLink) {
            return;
        }
        if (!this.hideIcon) {
            this.createIcon();
        }
    };
    /**
     * @private
     * @return {?}
     */
    ExternalLinkDirective.prototype.createIcon = /**
     * @private
     * @return {?}
     */
    function () {
        // tslint:disable-next-line:no-unused-expression
        this.vc.constructor.name === "ViewContainerRef_";
        /** @type {?} */
        var factory = this.cfr.resolveComponentFactory(FaIconComponent);
        /** @type {?} */
        var component = this.vc.createComponent(factory);
        component.instance.iconProp = ['fas', 'external-link-alt'];
        /** @type {?} */
        var spanElement = document.createElement('span');
        spanElement.classList.add('margin-left-2px');
        /** @type {?} */
        var supElement = document.createElement('sup');
        supElement.appendChild(component.location.nativeElement);
        spanElement.appendChild(supElement);
        this.el.nativeElement.appendChild(spanElement);
        component.instance.ngOnChanges({});
    };
    ExternalLinkDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'a[href]'
                },] }
    ];
    /** @nocollapse */
    ExternalLinkDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ComponentFactoryResolver },
        { type: ViewContainerRef }
    ]; };
    ExternalLinkDirective.propDecorators = {
        href: [{ type: Input }],
        hideIcon: [{ type: Input }]
    };
    return ExternalLinkDirective;
}());
export { ExternalLinkDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ExternalLinkDirective.prototype.vcRef;
    /** @type {?} */
    ExternalLinkDirective.prototype.href;
    /** @type {?} */
    ExternalLinkDirective.prototype.hideIcon;
    /**
     * @type {?}
     * @private
     */
    ExternalLinkDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    ExternalLinkDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    ExternalLinkDirective.prototype.cfr;
    /**
     * @type {?}
     * @private
     */
    ExternalLinkDirective.prototype.vc;
}
