// @gsa-sam/components/esm5/lib/search-result-list/search-result-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ContentChild, TemplateRef } from '@angular/core';
var SdsSearchResultListComponent = /** @class */ (function () {
    function SdsSearchResultListComponent() {
        /**
         * Show divider between results
         */
        this.divider = true;
    }
    SdsSearchResultListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-search-result-list',
                    template: "<div *ngFor=\"let item of model\">\n    <hr *ngIf=\"divider\" class=\"thin\" />\n    <ng-container *ngTemplateOutlet=\"resultContentTemplate,\n                   context: { $implicit: item }\"></ng-container>\n</div>\n<hr *ngIf=\"divider && model?.length\" class=\"thin\" />\n<div *ngIf=\"!model|| model.length===0\">\n    <div class=\"grid-row border-1px border-base-lighter bg-base-lightest\">\n      <div class=\"grid-col-auto text-center text-base margin-4\">\n        <fa-icon [icon]=\"['sds', 'search']\" [classes]=\"['search']\" size=\"7x\"></fa-icon>\n      </div>\n      <div class=\"grid-col-fill display-flex flex-column padding-y-1\">\n        <h1 class=\"margin-bottom-105 text-semibold\">No matches found</h1>\n        <p class=\"margin-y-0 font-sans-lg\">We couldn't find a match for your search criteria.</p>\n        <p class=\"margin-y-0 font-sans-lg\">Please try another search or go back to previous results.</p>\n        <button class=\"usa-button usa-button--secondary width-card margin-y-2\">Go Back</button>\n      </div>\n    </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    SdsSearchResultListComponent.propDecorators = {
        model: [{ type: Input }],
        divider: [{ type: Input }],
        resultContentTemplate: [{ type: ContentChild, args: ['resultContent',] }]
    };
    return SdsSearchResultListComponent;
}());
export { SdsSearchResultListComponent };
if (false) {
    /**
     * List of items
     * @type {?}
     */
    SdsSearchResultListComponent.prototype.model;
    /**
     * Show divider between results
     * @type {?}
     */
    SdsSearchResultListComponent.prototype.divider;
    /**
     * Child Template to be used to display the data for each item in the list of items
     * @type {?}
     */
    SdsSearchResultListComponent.prototype.resultContentTemplate;
}
