// @gsa-sam/components/esm5/lib/text/child.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
var SdsTextChildComponent = /** @class */ (function () {
    function SdsTextChildComponent() {
        this.itemsChange = new EventEmitter();
    }
    // Method to remove an item from the items array, emits an event that the parent component is listening for
    // Method to remove an item from the items array, emits an event that the parent component is listening for
    /**
     * @param {?} index
     * @return {?}
     */
    SdsTextChildComponent.prototype.removeItem = 
    // Method to remove an item from the items array, emits an event that the parent component is listening for
    /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.items.splice(index, 1);
        this.itemsChange.emit(this.items);
    };
    SdsTextChildComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-text-child',
                    template: "\n    <ng-container *ngIf=\"items\">\n      <span class=\"sds-tag sds-tag--chip margin-x-05\" style=\"cursor:pointer;\" *ngFor=\"let item of items; let i=index\" (click)=\"removeItem(i)\">{{item}} <small class=\"margin-left-05\">(x)</small></span>\n    </ng-container>\n  "
                }] }
    ];
    SdsTextChildComponent.propDecorators = {
        items: [{ type: Input }],
        itemsChange: [{ type: Output }]
    };
    return SdsTextChildComponent;
}());
export { SdsTextChildComponent };
if (false) {
    /** @type {?} */
    SdsTextChildComponent.prototype.items;
    /** @type {?} */
    SdsTextChildComponent.prototype.itemsChange;
}
