// @gsa-sam/components/esm5/lib/text/text.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var SdsTextComponent = /** @class */ (function () {
    function SdsTextComponent(cd) {
        this.cd = cd;
        this.items = [];
        this.multiple = true;
        this._onChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { });
        this._onTouched = (/**
         * @return {?}
         */
        function () { });
    }
    // Helper method to programatically add a value to the existing items array
    // Helper method to programatically add a value to the existing items array
    /**
     * @param {?} val
     * @return {?}
     */
    SdsTextComponent.prototype.addItem = 
    // Helper method to programatically add a value to the existing items array
    /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        if (this.multiple) {
            this.items = tslib_1.__spread(this.items, [val]);
            this.updateModel();
        }
    };
    // Method that is fired when the child component event notifies us that the items array has been modified within the child component
    // Method that is fired when the child component event notifies us that the items array has been modified within the child component
    /**
     * @param {?} $event
     * @return {?}
     */
    SdsTextComponent.prototype.updateItems = 
    // Method that is fired when the child component event notifies us that the items array has been modified within the child component
    /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.updateModel();
    };
    // Helper method that gets a new instance of the model and notifies ControlValueAccessor that we have a new model for this FormControl (our custom component)
    // Helper method that gets a new instance of the model and notifies ControlValueAccessor that we have a new model for this FormControl (our custom component)
    /**
     * @return {?}
     */
    SdsTextComponent.prototype.updateModel = 
    // Helper method that gets a new instance of the model and notifies ControlValueAccessor that we have a new model for this FormControl (our custom component)
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var model = this.getModel();
        this._onChange(model);
    };
    // Helper method to return a new instance of an array that contains our items
    // Helper method to return a new instance of an array that contains our items
    /**
     * @return {?}
     */
    SdsTextComponent.prototype.getModel = 
    // Helper method to return a new instance of an array that contains our items
    /**
     * @return {?}
     */
    function () {
        return tslib_1.__spread(this.items);
    };
    // ControlValueAccessor (and Formly) is trying to update the value of the FormControl (our custom component) programatically
    // If there is a value we will just overwrite items
    // If there is no value we reset the items array to be empty
    // ControlValueAccessor (and Formly) is trying to update the value of the FormControl (our custom component) programatically
    // If there is a value we will just overwrite items
    // If there is no value we reset the items array to be empty
    /**
     * @param {?} value
     * @return {?}
     */
    SdsTextComponent.prototype.writeValue = 
    // ControlValueAccessor (and Formly) is trying to update the value of the FormControl (our custom component) programatically
    // If there is a value we will just overwrite items
    // If there is no value we reset the items array to be empty
    /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value && value.length && this.items !== value) {
            this.items = value;
            this.cd.markForCheck();
        }
        else {
            this.items = [];
            this.cd.markForCheck();
        }
    };
    // ControlValueAccessor hook that lets us call this._onChange(var) to let the form know our variable has changed (in this case model)
    // ControlValueAccessor hook that lets us call this._onChange(var) to let the form know our variable has changed (in this case model)
    /**
     * @param {?} fn
     * @return {?}
     */
    SdsTextComponent.prototype.registerOnChange = 
    // ControlValueAccessor hook that lets us call this._onChange(var) to let the form know our variable has changed (in this case model)
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onChange = fn;
    };
    // ControlValueAccessor hook (not used)
    // ControlValueAccessor hook (not used)
    /**
     * @param {?} fn
     * @return {?}
     */
    SdsTextComponent.prototype.registerOnTouched = 
    // ControlValueAccessor hook (not used)
    /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onTouched = fn;
    };
    SdsTextComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-text',
                    template: "\n    <div>\n      <input #searchInput class=\"usa-input display-inline-block\" />\n      <button class=\"usa-button margin-left-05 display-inline-block\" (click)=\"addItem(searchInput.value); searchInput.value=''\">Add Item</button>\n    </div>\n\n    <h4>Component Items</h4>\n    <pre>{{ items | json }}</pre>\n\n    <hr />\n\n    <h4>Child Component Items <small>(click to remove)</small></h4>\n    <sds-text-child [(items)]=\"items\" (itemsChange)=\"updateItems($event)\"></sds-text-child>\n  ",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return SdsTextComponent; })),
                            multi: true
                        }
                    ],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    SdsTextComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    return SdsTextComponent;
}());
export { SdsTextComponent };
if (false) {
    /** @type {?} */
    SdsTextComponent.prototype.items;
    /** @type {?} */
    SdsTextComponent.prototype.multiple;
    /**
     * @type {?}
     * @private
     */
    SdsTextComponent.prototype._onChange;
    /**
     * @type {?}
     * @private
     */
    SdsTextComponent.prototype._onTouched;
    /**
     * @type {?}
     * @private
     */
    SdsTextComponent.prototype.cd;
}
