// @gsa-sam/components/esm5/lib/autocomplete-search/models/SDSAutocompleteConfiguration.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { SelectionMode } from '../../selected-result/models/sds-selected-item-model-helper';
var SDSAutocompleteSearchConfiguration = /** @class */ (function () {
    /**
     * sets the default debounce time to 250 milliseconds
     */
    function SDSAutocompleteSearchConfiguration() {
        /**
         * Mode of the model either allows a single item or multiple items
         */
        this.selectionMode = SelectionMode.SINGLE;
        /**
         * Allows option to allow user text not in the standard results
         */
        this.isFreeTextEnabled = false;
        /**
         * Text appeneded ad the end of free text
         */
        this.freeTextSubtext = 'search';
        /**
         * Focus into autocomplete search
         */
        this.focusInSearch = true;
        /**
         * The aria-label for the auto-complete
         */
        this.ariaLabelText = 'Auto Complete';
        /**
         * To enable the tag mode
         */
        this.isTagModeEnabled = false;
        /**
         * To make input readonly
         */
        this.inputReadOnly = false;
        this.debounceTime = 250;
        this.minimumCharacterCountSearch = 0;
    }
    return SDSAutocompleteSearchConfiguration;
}());
export { SDSAutocompleteSearchConfiguration };
if (false) {
    /**
     * Used to describe the drop down (Text should match the label that will be supplied)
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.labelText;
    /**
     * Used for the Id of the control
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.id;
    /**
     *  This is the primary field used to identify each object in the results
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.primaryKeyField;
    /**
     *  Property from supplied model used for the top part of the basic template
     *  and the text for single selection
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.primaryTextField;
    /**
     *  Property from supplied model used for the bottom part of the basic template
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.secondaryTextField;
    /**
     *  Sets the time waited for addional key actions Default is 250 milliseconds
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.debounceTime;
    /**
     * Place holder text for autocomplete input
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.autocompletePlaceHolderText;
    /**
     * Mininumn Characters for search
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.minimumCharacterCountSearch;
    /**
     * Mode of the model either allows a single item or multiple items
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.selectionMode;
    /**
     * Allows option to allow user text not in the standard results
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.isFreeTextEnabled;
    /**
     * Text appeneded ad the end of free text
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.freeTextSubtext;
    /**
     * Focus into autocomplete search
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.focusInSearch;
    /**
     * The aria-label for the auto-complete
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.ariaLabelText;
    /**
     * To enable the tag mode
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.isTagModeEnabled;
    /**
     * To make input readonly
     * @type {?}
     */
    SDSAutocompleteSearchConfiguration.prototype.inputReadOnly;
}
