// @gsa-sam/components/esm2015/lib/selected-result/selected-result.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef, forwardRef } from '@angular/core';
import { SDSSelectedItemModel } from './models/sds-selectedItem.model';
import { SDSSelectedResultConfiguration } from './models/SDSSelectedResultConfiguration';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { SDSSelectedItemModelHelper } from './models/sds-selected-item-model-helper';
/** @type {?} */
const SDS_SelectedResult_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    () => SDSSelectedResultComponent)),
    multi: true
};
export class SDSSelectedResultComponent {
    constructor() {
        /**
         * Stored Event for ControlValueAccessor
         */
        this.onTouchedCallback = (/**
         * @return {?}
         */
        () => null);
        /**
         * Stored Event for ControlValueAccessor
         */
        this.propogateChange = (/**
         * @param {?} _
         * @return {?}
         */
        (_) => null);
    }
    /**
     * Removes item from the model
     * @param {?} item
     * @return {?}
     */
    removeItem(item) {
        if (!this.disabled) {
            SDSSelectedItemModelHelper.removeItem(item, this.configuration.primaryKeyField, this.model.items);
            this.propogateChange(this.model);
            this.onTouchedCallback();
        }
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    writeValue(obj) {
        if (obj instanceof SDSSelectedItemModel) {
            this.model = (/** @type {?} */ (obj));
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.propogateChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * Gets the string value from the specifed properties of an object
     * @param {?} object
     * @param {?} propertyFields comma seperated list with periods depth of object
     * @return {?}
     */
    getObjectValue(object, propertyFields) {
        /** @type {?} */
        let value = '';
        /** @type {?} */
        let current = object;
        /** @type {?} */
        let fieldSplit = propertyFields.split(',');
        for (let i = 0; i < fieldSplit.length; i++) {
            /** @type {?} */
            let fieldValue = fieldSplit[i];
            /** @type {?} */
            let fieldPartSplit = fieldValue.split('.');
            for (let j = 0; j < fieldPartSplit.length; j++) {
                /** @type {?} */
                let fieldCheckValue = fieldPartSplit[j];
                if (current) {
                    current = current[fieldCheckValue];
                }
            }
            if (current) {
                value += current.toString() + ' ';
            }
            current = object;
        }
        return value.trim();
    }
}
SDSSelectedResultComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-selected-result',
                template: "<ul  *ngIf=\"model\" [attr.aria-label]=\"configuration.labelText +' results'\" class=\"usa-list usa-list--unstyled sds-autocomplete-selected\" aria-relevant=\"additions\" role=\"listbox\" aria-live=\"polite\">\n    <li role=\"option\" *ngFor=\"let result of model.items; let i = index\">\n        <div [attr.class]=\" disabled ? 'sds-tag sds-tag--chip sds-autocomplete-selected__item sds-autocomplete-selected__item--disabled' :'sds-tag sds-tag--chip sds-autocomplete-selected__item'\">\n            <ng-container *ngIf=\"itemTemplate\" [ngTemplateOutlet]=\"itemTemplate\" [ngTemplateOutletContext]=\"{$implicit:result}\">\n            </ng-container>\n            <ng-container *ngIf=\"!itemTemplate\">\n                <span class=\"display-block text-semibold\"> {{ getObjectValue(result, configuration.primaryTextField) }}\n                </span>\n                <ng-container *ngIf=\"configuration.secondaryTextField && result[configuration.secondaryTextField] \">\n                    {{ result[configuration.secondaryTextField] }}\n                </ng-container>\n            </ng-container>\n            <button [attr.aria-label]=\"'Remove Item ' + getObjectValue(result, configuration.primaryTextField)\" [class.text-base]=\"disabled\" aria-hidden=\"false\" class=\"sds-tag__close  position-absolute right-1 top-2px\" (click)=\"removeItem(result)\" (keyup.enter)=\"removeItem(result)\">\n        <fa-icon [icon]=\"['fas', 'times']\" size=\"sm\"></fa-icon>\n      </button>\n        </div>\n    </li>\n</ul>",
                providers: [SDS_SelectedResult_VALUE_ACCESSOR],
                styles: [""]
            }] }
];
SDSSelectedResultComponent.propDecorators = {
    itemTemplate: [{ type: Input }],
    configuration: [{ type: Input }],
    disabled: [{ type: Input }]
};
if (false) {
    /**
     * Allow to insert a customized template for suggestions to use
     * @type {?}
     */
    SDSSelectedResultComponent.prototype.itemTemplate;
    /**
     * The data model that has the selected item
     * @type {?}
     */
    SDSSelectedResultComponent.prototype.model;
    /**
     * Configuration for the Selected Results control
     * @type {?}
     */
    SDSSelectedResultComponent.prototype.configuration;
    /**
     * Stored Event for ControlValueAccessor
     * @type {?}
     */
    SDSSelectedResultComponent.prototype.onTouchedCallback;
    /**
     * Stored Event for ControlValueAccessor
     * @type {?}
     */
    SDSSelectedResultComponent.prototype.propogateChange;
    /** @type {?} */
    SDSSelectedResultComponent.prototype.disabled;
}
