// @gsa-sam/components/esm5/lib/top-banner/top-banner.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var SdsTopBannerComponent = /** @class */ (function () {
    function SdsTopBannerComponent() {
        this.showDetail = false;
        this.description = '';
    }
    /**
     * @return {?}
     */
    SdsTopBannerComponent.prototype.toggleDetails = /**
     * @return {?}
     */
    function () {
        this.showDetail = !this.showDetail;
    };
    /**
     * @return {?}
     */
    SdsTopBannerComponent.prototype.closeDetail = /**
     * @return {?}
     */
    function () {
        if (this.showDetail) {
            this.showDetail = false;
        }
    };
    SdsTopBannerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sds-top-banner',
                    template: "<div class=\"usa-banner\">\n  <div class=\"usa-accordion\">\n    <header\n      class=\"usa-banner__header\"\n      [class.sam-banner__header--expanded]=\"showDetail\"\n    >\n      <div class=\"usa-banner__inner\">\n        <div class=\"grid-col-auto\">\n          <span class=\"usa-banner__header-flag\"></span>\n        </div>\n        <div class=\"grid-col-fill tablet:grid-col-auto\">\n          <p class=\"usa-banner__header-text\">\n            An official website of the United States government\n          </p>\n          <p class=\"usa-banner__header-action\" aria-hidden=\"true\">\n            Here\u2019s how you know\n          </p>\n        </div>\n        <button\n          class=\"usa-accordion__button usa-banner__button\"\n          (click)=\"toggleDetails()\"\n          (blur)=\"closeDetail()\"\n          [attr.aria-expanded]=\"showDetail\"\n          aria-controls=\"gov-banner\"\n        >\n          <span class=\"usa-banner__button-text\">Here\u2019s how you know</span>\n        </button>\n        <div class=\"usa-banner__header-description\">\n          <em>{{ description }}</em>\n        </div>\n      </div>\n    </header>\n    <div\n      class=\"usa-banner__content usa-accordion__content\"\n      id=\"gov-banner\"\n      [hidden]=\"!showDetail\"\n    >\n      <div class=\"grid-row grid-gap-lg\">\n        <div class=\"usa-banner__guidance tablet:grid-col-6\">\n          <span class=\"usa-banner__icon usa-media-block__img\"></span>\n          <div class=\"usa-media-block__body\">\n            <p>\n              <strong>The .gov means it\u2019s official.</strong> <br />\n              Federal government websites often end in .gov or .mil. Before\n              sharing sensitive information, make sure you\u2019re on a federal\n              government site.\n            </p>\n          </div>\n        </div>\n        <div class=\"usa-banner__guidance tablet:grid-col-6\">\n          <span class=\"usa-banner__icon usa-media-block__img\"></span>\n          <div class=\"usa-media-block__body\">\n            <p>\n              <strong>The site is secure.</strong> <br />\n              The <strong>https://</strong> ensures that you are connecting to\n              the official website and that any information you provide is\n              encrypted and transmitted securely.\n            </p>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n",
                    styles: [""]
                }] }
    ];
    SdsTopBannerComponent.propDecorators = {
        description: [{ type: Input }]
    };
    return SdsTopBannerComponent;
}());
export { SdsTopBannerComponent };
if (false) {
    /** @type {?} */
    SdsTopBannerComponent.prototype.showDetail;
    /** @type {?} */
    SdsTopBannerComponent.prototype.description;
}
