// @gsa-sam/components/esm2015/lib/pagination/pagination.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output, ChangeDetectorRef, ViewChild, ElementRef } from '@angular/core';
import { PaginationConfigurationModel, PaginationModel } from './model/paginationModel';
export class PaginationComponent {
    /**
     * @param {?} change
     */
    constructor(change) {
        this.change = change;
        /**
         * Output of the page model object
         */
        this.pageChange = new EventEmitter();
        /**
         * Pagination model
         */
        this.page = new PaginationModel();
        /**
         * debounce time for current page input
         */
        this.debounceTime = 500;
        /**
         * Drop down options for page size
         */
        this.options = [
            { label: '25', value: 25 },
            { label: '50', value: 50 },
            { label: '100', value: 100 }
        ];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.maintainPreviousValue();
    }
    /**
     * Stores the previous number. Used when focus out if field empty
     * @private
     * @return {?}
     */
    maintainPreviousValue() {
        this.previousNumber = this.page.pageNumber.valueOf();
    }
    /**
     * previous page lowers page number by one within range
     * @return {?}
     */
    previousPage() {
        if (this.page.pageNumber > 1) {
            this.page.pageNumber--;
            this.maintainPreviousValue();
            this.pageChange.emit(this.page);
        }
    }
    /**
     * next page increase page number by one within range
     * @return {?}
     */
    nextPage() {
        if (this.page.pageNumber < this.page.totalPages) {
            this.page.pageNumber++;
            this.maintainPreviousValue();
            this.pageChange.emit(this.page);
        }
    }
    /**
     * current page changes sets new value if within range
     * @param {?=} newValue
     * @return {?}
     */
    valuechange(newValue) {
        window.clearTimeout(this.timeoutNumber);
        this.timeoutNumber = window.setTimeout((/**
         * @return {?}
         */
        () => {
            if (newValue || newValue === 0) {
                newValue = this.handleInputOutsideBounds(newValue);
                if (newValue >= 1 && newValue <= this.page.totalPages) {
                    this.page.pageNumber = newValue;
                    this.maintainPreviousValue();
                    this.pageChange.emit(this.page);
                    this.change.detectChanges();
                }
            }
            else {
                if (this.page.pageNumber) {
                    this.maintainPreviousValue();
                }
            }
        }), this.debounceTime);
    }
    /**
     * adjusts the value if not within the page limit above or below
     * @private
     * @param {?=} newValue handles
     * @return {?}
     */
    handleInputOutsideBounds(newValue) {
        if (newValue < 1) {
            newValue = 1;
            this.currentPageField.nativeElement.value = newValue;
        }
        else if (newValue > this.page.totalPages) {
            newValue = this.page.totalPages;
            this.currentPageField.nativeElement.value = newValue;
        }
        return newValue;
    }
    /**
     * current page focus out will replace with previous valid if empty
     * @return {?}
     */
    currentPageFocusOut() {
        if (this.currentPageField.nativeElement.value === '') {
            this.currentPageField.nativeElement.value = this.page.pageNumber = this.previousNumber;
            this.change.detectChanges();
        }
    }
    /**
     * page size selection change
     * @return {?}
     */
    onSelectChange() {
        this.page.pageNumber = 1;
        this.pageChange.emit(this.page);
    }
}
PaginationComponent.decorators = [
    { type: Component, args: [{
                selector: 'sds-pagination',
                template: "<div class=\"sds-pagination\">\n  <div class=\"sds-pagination__controls\">\n    <button [attr.id]=\"paginationConfiguration.id +'-previousPage'\" (click)=\"previousPage()\">Previous Page</button>\n    <label class=\"usa-label font-sans-3xs margin-top-1 text-italic text-base usa-sr-only\"\n      [attr.for]=\"paginationConfiguration.id + '-currentPage'\">Current Page</label>\n    <input [attr.id]=\"paginationConfiguration.id +'-currentPage'\"\n      class=\"usa-input height-3 width-4 line-height-05 font-sans-3xs text-center border-base-light\" #currentPage\n      (ngModelChange)=\"valuechange($event)\" [(ngModel)]=\"page.pageNumber\" type=\"number\" min=\"1\"\n      [(attr.max)]=\"page.totalPages\" (focusout)=\"currentPageFocusOut()\" aria-label=\"current page\" />\n    <span class=\"sds-pagination__total\">\n      of <strong>{{ page.totalPages }}</strong>\n    </span>\n    <button [attr.id]=\"paginationConfiguration.id +'-nextPage'\" (click)=\"nextPage()\">Next Page</button>\n  </div>\n  <div class=\"sds-pagination__results\">\n    <label class=\"usa-label font-sans-3xs text-italic text-base\"\n      [attr.for]=\"paginationConfiguration.id + '-select'\">Results per page</label>\n    <select class=\"usa-select usa-select--small border-base-light\" [attr.id]=\"paginationConfiguration.id + '-select'\"\n      (change)=\"onSelectChange()\" [(ngModel)]=\"page.pageSize\">\n      <option *ngFor=\"let item of options\" [ngValue]=\"item.value\">\n        {{ item.label }}\n      </option>\n    </select>\n  </div>\n</div>\n",
                styles: ["input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}input[type=number]{-moz-appearance:textfield}"]
            }] }
];
/** @nocollapse */
PaginationComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PaginationComponent.propDecorators = {
    currentPageField: [{ type: ViewChild, args: ['currentPage',] }],
    pageChange: [{ type: Output }],
    page: [{ type: Input }],
    paginationConfiguration: [{ type: Input }]
};
if (false) {
    /**
     * Input field for the current page
     * @type {?}
     */
    PaginationComponent.prototype.currentPageField;
    /**
     * Output of the page model object
     * @type {?}
     */
    PaginationComponent.prototype.pageChange;
    /**
     * Pagination model
     * @type {?}
     */
    PaginationComponent.prototype.page;
    /**
     * configuration for the pagination
     * @type {?}
     */
    PaginationComponent.prototype.paginationConfiguration;
    /**
     * debounce time for current page input
     * @type {?}
     */
    PaginationComponent.prototype.debounceTime;
    /**
     * Stores the previous number
     * @type {?}
     * @private
     */
    PaginationComponent.prototype.previousNumber;
    /**
     * timeout id of the debounce time
     * @type {?}
     */
    PaginationComponent.prototype.timeoutNumber;
    /**
     * Drop down options for page size
     * @type {?}
     */
    PaginationComponent.prototype.options;
    /**
     * @type {?}
     * @private
     */
    PaginationComponent.prototype.change;
}
