// @gsa-sam/components/esm5/lib/accordion/accordion.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Input, ContentChildren, QueryList } from "@angular/core";
import { CdkAccordion } from "@angular/cdk/accordion";
import { FocusKeyManager } from "@angular/cdk/a11y";
import { HOME, END } from "@angular/cdk/keycodes";
import { SDS_ACCORDION } from "./accordion-base";
import { SdsAccordionItemHeaderComponent } from "./accordion-item-header.component";
var SdsAccordionDirective = /** @class */ (function (_super) {
    tslib_1.__extends(SdsAccordionDirective, _super);
    function SdsAccordionDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.displayMode = "default";
        return _this;
    }
    /**
     * @return {?}
     */
    SdsAccordionDirective.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this._keyManager = new FocusKeyManager(this._headers).withWrap();
    };
    /** Handles keyboard events coming in from the item headers. */
    /**
     * Handles keyboard events coming in from the item headers.
     * @param {?} event
     * @return {?}
     */
    SdsAccordionDirective.prototype._handleHeaderKeydown = /**
     * Handles keyboard events coming in from the item headers.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var keyCode = event.keyCode;
        if (keyCode === HOME) {
            this._keyManager.setFirstItemActive();
            event.preventDefault();
        }
        else if (keyCode === END) {
            this._keyManager.setLastItemActive();
            event.preventDefault();
        }
        else {
            this._keyManager.onKeydown(event);
        }
    };
    /**
     * @param {?} header
     * @return {?}
     */
    SdsAccordionDirective.prototype._handleHeaderFocus = /**
     * @param {?} header
     * @return {?}
     */
    function (header) {
        this._keyManager.updateActiveItem(header);
    };
    SdsAccordionDirective.decorators = [
        { type: Directive, args: [{
                    selector: "sds-accordion",
                    exportAs: "sdsAccordion",
                    inputs: ["multi"],
                    providers: [
                        {
                            provide: SDS_ACCORDION,
                            useExisting: SdsAccordionDirective
                        }
                    ],
                    host: {
                        'class': 'sds-accordion',
                        '[class.sds-accordion--basic]': 'displayMode === "basic"',
                    }
                },] }
    ];
    SdsAccordionDirective.propDecorators = {
        _headers: [{ type: ContentChildren, args: [SdsAccordionItemHeaderComponent, { descendants: true },] }],
        displayMode: [{ type: Input }]
    };
    return SdsAccordionDirective;
}(CdkAccordion));
export { SdsAccordionDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SdsAccordionDirective.prototype._keyManager;
    /** @type {?} */
    SdsAccordionDirective.prototype._headers;
    /** @type {?} */
    SdsAccordionDirective.prototype.displayMode;
}
