// @gsa-sam/layouts/esm5/lib/feature/search-list-layout/model/search-list-layout.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SearchParameters = /** @class */ (function () {
    function SearchParameters() {
    }
    return SearchParameters;
}());
export { SearchParameters };
if (false) {
    /**
     * page
     * @type {?}
     */
    SearchParameters.prototype.page;
    /**
     * Sort value
     * @type {?}
     */
    SearchParameters.prototype.sortField;
    /**
     * filter data
     * @type {?}
     */
    SearchParameters.prototype.filter;
}
/**
 * @record
 */
export function Page() { }
if (false) {
    /**
     * Page number
     * @type {?}
     */
    Page.prototype.pageNumber;
    /**
     * Page size
     * @type {?}
     */
    Page.prototype.pageSize;
    /**
     * Total number of pages
     * @type {?}
     */
    Page.prototype.totalPages;
}
var SearchResult = /** @class */ (function () {
    function SearchResult() {
    }
    return SearchResult;
}());
export { SearchResult };
if (false) {
    /**
     * Total number of items (beyond the page)
     * @type {?}
     */
    SearchResult.prototype.totalItems;
    /**
     * Items to be displayed
     * @type {?}
     */
    SearchResult.prototype.items;
}
/**
 * @record
 */
export function SearchListInterface() { }
if (false) {
    /**
     * Method to get the takes it takes in the SearchParameters and returns SearchResult object
     * @param {?} search
     * @return {?}
     */
    SearchListInterface.prototype.getData = function (search) { };
}
var sortItem = /** @class */ (function () {
    function sortItem() {
    }
    return sortItem;
}());
export { sortItem };
if (false) {
    /**
     * Text to be displayed
     * @type {?}
     */
    sortItem.prototype.text;
    /**
     * Value of Item
     * @type {?}
     */
    sortItem.prototype.value;
}
var SearchListConfiguration = /** @class */ (function () {
    function SearchListConfiguration() {
        /**
         * Starting page size
         */
        this.pageSize = 25;
    }
    return SearchListConfiguration;
}());
export { SearchListConfiguration };
if (false) {
    /**
     * List of sort by items
     * @type {?}
     */
    SearchListConfiguration.prototype.sortList;
    /**
     * default sort value
     * @type {?}
     */
    SearchListConfiguration.prototype.defaultSortValue;
    /**
     * Starting page size
     * @type {?}
     */
    SearchListConfiguration.prototype.pageSize;
}
